---
title: "Differential Methylated Regions Functional Enrichment Analyses"
author: "Hamish Williams"
date: "`r Sys.Date()`"
output: word_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Overview

Performing over representation analysis on the DMR gene lists for both experiments to determine if there is some statistcially enriched terms and functions. This analysis is only going to contribute to highlighting if particular functions are being enriched, and not to remove genes associated to DMRs away from the results and discussion. The results of the DMR anlaysis have already gone through a lengthy and rigouros bout of statistical analyses to get to a final list of DMRs, hence this analysis is only to highlight results which may be of greater importance to the physiological response of actinia equina to their experimental contexts.

# Installing and Loading required packages

```{r check_packages_ClusterProfilerR, message=FALSE, warning=FALSE, results='hide'}
# Build a function which will automatically detect installed packages and install required packages
check_packages_ClusterProfileR <- function(pkg_list) {
  for (pkg in pkg_list) {
    if (!require(pkg, character.only = TRUE)) {
      print(paste(pkg, "is not installed, installing package..."))
      install.packages(pkg) # installs regular R packages
      BiocManager::install(pkg) # installs packages requiring BiocManager
    } else {
      suppressMessages(library(pkg, character.only = TRUE))
      print(paste(pkg, "is installed and loaded"))
    }
  }
}

# Generate a dataframe with a list of packages required

pkg_list_ClusterProfileR<-c("BiocManager", # Have this first to ensure BiocManager packages will be installed
                            "tidyverse", 
                            "tibble", 
                            "stats",
                            "ggplot2", 
                            "dplyr", 
                            "tidyr", 
                            "ggfortify", 
                            "knitr",
                            "clusterProfiler",
                            "AnnotationForge",
                            "enrichplot",
                            "data.table",
                            "stats",
                            "plotly",
                            "ggVennDiagram",
                            "VennDiagram",
                            "limma",
                            "ggridges",
                            "pathview",
                            "KEGGREST",
                            "stringr",
                            "gridExtra",
                            "ggside",
                            "gghalves",
                            "flextable",
                            "officer",
                            "ggh4x"
                            )


# Run the code
  # NOTE: this code should be run twice, double checking is always good!

# Run Function::
check_packages_ClusterProfileR(pkg_list_ClusterProfileR)
# You will need to answer "no" when asked if you want to update packages, otherwise it will take a while to load

# Double Check everything has loaded:
check_packages_ClusterProfileR(pkg_list_ClusterProfileR)
```

## convertFractionToNumeric

Used to convert gene ratios in final analyses to numerical values for generating figures

```{r convertFractionToNumeric, message=FALSE, warning=FALSE}
convertFractionToNumeric <- function(fraction) {
  fraction_parts <- strsplit(fraction, "/")[[1]]
  numerator <- as.numeric(fraction_parts[1])
  denominator <- as.numeric(fraction_parts[2])
  numeric_value <- numerator / denominator
  return(numeric_value)
}


# Make function to capture methylation data for each gene in the GSEA results
get_methDiff_data <- function(core_enrichment, meth_data) {
  genes <- unlist(strsplit(core_enrichment, "/"))
  lfc_values <- meth_data$meth.diff[match(genes, meth_data$gene_id_labelling)]
  paste(lfc_values, collapse = "/")
}
```

# Importing Data:

```{r import_data}
getwd()
setwd("C:/Users/r02hw22/OneDrive - University of Aberdeen/Documents/GitHub/Methylation_Analyses/Methylation_Analyses/All context DMR results")
getwd()


# load in the gene lists for experiments 1 & 2;
exp1_DMR_list <- read.table("Exp1_DMRs_gene_id_matched.txt", header = T, sep = '\t')
exp1_DMR_list <- subset(exp1_DMR_list[exp1_DMR_list$p_fdr <= 0.1,])
exp1_DMR_list<- as.vector(exp1_DMR_list[,19])
exp2_DMR_list <- read.table("Exp2_DMRs_gene_id_matched.txt", header = T, sep = '\t')
exp2_DMR_list <- subset(exp2_DMR_list[exp2_DMR_list$p_fdr <= 0.1,])
exp2_DMR_list<- as.vector(exp2_DMR_list[,19])


# Genome Matching file:
matching_file <- read.table("genome_matching_file.txt", 
                            header = F, 
                            sep = '\t', 
                            stringsAsFactors = FALSE)

names(matching_file)[names(matching_file) == "V13"] <- "ID"

# Reduce list of hits per gene to only the best per gene id:
matching_file_best_hits <- matching_file %>%
  group_by(V1) %>%
  filter(V11 == min(V11)) %>%
  ungroup()

matching_file_best_hits <- unique(matching_file_best_hits)

# Load in uniprot data
BP_MF_Terms <- read.table("UNIPROT_results.txt", 
                          header = T, 
                          sep = '\t', 
                          stringsAsFactors = FALSE)

names(BP_MF_Terms)[names(BP_MF_Terms) == "From"] <- "ID"
names(BP_MF_Terms)[names(BP_MF_Terms) == "Gene.Names"] <- "GENENAMES"
names(BP_MF_Terms)[names(BP_MF_Terms) == "Gene.Ontology..biological.process."] <- "GO:BP"
names(BP_MF_Terms)[names(BP_MF_Terms) == "Gene.Ontology..molecular.function."] <- "GO:MF"

```

## Merge Blastp and GO terms files together

```{r Merge_Files}
# Merge DEG Files:
merged_df <- merge(matching_file_best_hits, BP_MF_Terms, by = "ID")

# Subset Dataframe
merged_df <- subset(merged_df[, c("V1","ID","GO:BP","GO:MF")])

# Change Variable names
names(merged_df)[names(merged_df) == "V1"] <- "GENE"

```

## Separate BP and MF GOs

```{r Separate_GOs}
# Select BP column:
BP <- merged_df[,c("GENE","ID","GO:BP")]
BP <- BP[!apply(BP == "", 1, any), ]
BP <- BP[!duplicated(BP), ]
# Extract GO terms:
GENES_BP <- BP
GENES_BP$GO <- ""
bracketed_rows <- grepl("\\[.*\\]", GENES_BP$`GO:BP`)
GENES_BP$GO[bracketed_rows] <- gsub(".*\\[(.*)\\].*", "\\1", GENES_BP$`GO:BP`[bracketed_rows])
GENES_BP <- GENES_BP[, c("GO","GENE")]
GENES_BP <- GENES_BP[!duplicated(GENES_BP), ]
GENES_BP <- GENES_BP[complete.cases(GENES_BP),]

## Extract Terms:
TERMs_BP <- BP
TERMs_BP$GO <- ""
bracketed_rows <- grepl("\\[.*\\]", TERMs_BP$`GO:BP`)
TERMs_BP$GO[bracketed_rows] <- gsub(".*\\[(.*)\\].*", "\\1", TERMs_BP$`GO:BP`[bracketed_rows])
TERMs_BP$`GO:BP`[bracketed_rows] <- gsub("\\[.*\\]", "", TERMs_BP$`GO:BP`[bracketed_rows])
TERMs_BP <- TERMs_BP[, c("GO","GO:BP")]
TERMs_BP <- TERMs_BP[!duplicated(TERMs_BP), ]
names(TERMs_BP)[names(TERMs_BP) == "GO:BP"] <- "GENENAMES_BP"
TERMs_BP <- TERMs_BP[complete.cases(TERMs_BP),]

# Select MF column:
MF <- merged_df[,c("GENE","ID","GO:MF")]
MF <- MF[!apply(MF == "", 1, any), ]
MF <- MF[!duplicated(MF), ]
# Extract GO terms:
GENES_MF <- MF
GENES_MF$GO <- ""
bracketed_rows <- grepl("\\[.*\\]", GENES_MF$`GO:MF`)
GENES_MF$GO[bracketed_rows] <- gsub(".*\\[(.*)\\].*", "\\1", GENES_MF$`GO:MF`[bracketed_rows])
GENES_MF <- GENES_MF[, c("GO","GENE")]
GENES_MF <- GENES_MF[!duplicated(GENES_MF), ]
GENES_MF <- GENES_MF[complete.cases(GENES_MF),]

## Extract Terms:
TERMs_MF <- MF
TERMs_MF$GO <- ""
bracketed_rows <- grepl("\\[.*\\]", TERMs_MF$`GO:MF`)
TERMs_MF$GO[bracketed_rows] <- gsub(".*\\[(.*)\\].*", "\\1", TERMs_MF$`GO:MF`[bracketed_rows])
TERMs_MF$`GO:MF`[bracketed_rows] <- gsub("\\[.*\\]", "", TERMs_MF$`GO:MF`[bracketed_rows])
TERMs_MF <- TERMs_MF[, c("GO","GO:MF")]
TERMs_MF <- TERMs_MF[!duplicated(TERMs_MF), ]
names(TERMs_MF)[names(TERMs_MF) == "GO:MF"] <- "GENENAMES_MF"
TERMs_MF <- TERMs_MF[complete.cases(TERMs_MF),]
```

# Creating loop to stratify the analysis per region type

```{r stratified_ORA_exp1}
# Load required packages
library(clusterProfiler)

# Load DMR data
dmr_data <- read.table("Exp1_DMRs_gene_id_matched.txt", header = TRUE, sep = '\t', stringsAsFactors = FALSE)

# Set the target region to analyze
  # promoter
  # three_prime_UTR
  # exon #results
  # gene #results
  # five_prime_UTR
  # downstream_region
  # dispersed_repeat #results
  # ncRNA_gene
{
target_region <- "ncRNA_gene"  # Change this to the region you want to analyze

# Subset data to selected region
region_data <- dmr_data[dmr_data$region == target_region, ]

# Filter for significant DMRs
sig_data <- region_data[region_data$p_fdr <= 0.1, ]

# keep the gene-context mapping from the filtered DMR data:
gene_context_map <- sig_data[!is.na(sig_data$gene_id_labelling) & sig_data$gene_id_labelling != "", 
                             c("gene_id_labelling", "context")]


# Extract gene IDs, removing NA or blank
gene_ids <- as.character(sig_data$gene_id_labelling)
gene_ids <- gene_ids[!is.na(gene_ids) & gene_ids != ""]

# Proceed only if there are valid gene IDs
if (length(gene_ids) > 0) {

  # Run ORA for Biological Processes (BP)
  ORA_BP <- enricher(gene = gene_ids,
                     TERM2GENE = GENES_BP,
                     TERM2NAME = TERMs_BP,
                     minGSSize = 1,
                     pvalueCutoff = 1,
                     qvalueCutoff = 1)

  # Run ORA for Molecular Functions (MF)
  ORA_MF <- enricher(gene = gene_ids,
                     TERM2GENE = GENES_MF,
                     TERM2NAME = TERMs_MF,
                     minGSSize = 1,
                     pvalueCutoff = 1,
                     qvalueCutoff = 1)

  # Extract results
  bp_results <- ORA_BP@result
    # If geneID is delimited (e.g., "GENE1/GENE2/GENE3"), split it
  bp_expanded <- bp_results %>%
    mutate(geneID = strsplit(as.character(geneID), "/")) %>%
    unnest(geneID)
  # Add context by joining
  bp_expanded <- left_join(bp_expanded, gene_context_map,
                           by = c("geneID" = "gene_id_labelling"),
                           relationship = "many-to-many")
  

  mf_results <- ORA_MF@result
      # If geneID is delimited (e.g., "GENE1/GENE2/GENE3"), split it
  mf_expanded <- mf_results %>%
    mutate(geneID = strsplit(as.character(geneID), "/")) %>%
    unnest(geneID)
  # Add context by joining
  mf_expanded <- left_join(mf_expanded, gene_context_map,
                           by = c("geneID" = "gene_id_labelling"),
                           relationship = "many-to-many")

  # Tag result source
  bp_expanded$GO <- "Biological Processes"
  mf_expanded$GO <- "Molecular Functions"

  # Combine and annotate results
  all_results <- rbind(bp_expanded, mf_expanded)
  all_results$region <- target_region
  all_results$experiment <- "Acute Experiment"

  # Convert ratios to numeric
  convertFractionToNumeric <- function(x) {
    parts <- strsplit(as.character(x), "/")[[1]]
    if (length(parts) == 2) {
      return(as.numeric(parts[1]) / as.numeric(parts[2]))
    } else {
      return(NA)
    }
  }

  all_results$GeneRatio <- sapply(all_results$GeneRatio, convertFractionToNumeric)
  all_results$BgRatio <- sapply(all_results$BgRatio, convertFractionToNumeric)

  # Save to CSV
  assign(paste0("ORA_results_", target_region, "_exp1"), all_results)
  output_file <- paste0("ORA_results_", target_region, ".csv")
  write.csv(all_results, output_file, row.names = FALSE)

  message("Analysis complete and results saved.")
} else {
  message("No significant gene IDs found for the selected region.")
}
}

# Combined all results together
ORA_results_exp1_combined <- rbind(ORA_results_dispersed_repeat_exp1,
                                   ORA_results_exon_exp1,
                                   ORA_results_gene_exp1)

# Plotting Data
ORA_results_exp1_combined <- ORA_results_exp1_combined[order(ORA_results_exp1_combined$GeneRatio), ]

ORA_results_exp1_combined_clean <- unique(ORA_results_exp1_combined)


ORA_exp1_dotplot <- ggplot(data = ORA_results_exp1_combined_clean, 
                              aes(y = reorder(Description, 
                                              Count, 
                                              increasing = T), 
                                  x = GeneRatio, 
                                  color = p.adjust)) + 
                       geom_point(aes(size = Count), 
                                  stroke = 4) +
                       labs(title = "Biological Processes", y = "GO Description", x = "Gene Ratio") +
                       theme_light() + 
                       scale_color_gradient(low = "red", high = "blue") + 
                       scale_size_continuous(breaks = seq(min(ORA_results_exp1_combined$Count), max(ORA_results_exp1_combined$Count), by = 1)) +
                       theme(
                         legend.spacing = unit(0.5, "cm"),         # Adjust overall legend spacing
                         legend.key.height = unit(3, "lines"), # Adjust the height between size examples
                         axis.text.y = element_text(size = 11)
                         ) + 
                       guides(size = guide_legend(title = "Count"))  + 
                       scale_y_discrete(labels = function(x) str_wrap(x, width = 28)) + # Adjust width for line breaks 
                       facet_wrap(~GO, scales = "free")

# Merge ORA_exp1_BP results and dotplot together, and rm dotplot
plot(ORA_exp1_dotplot)

## Export Dataframe results
write.csv(ORA_results_exp1_combined_clean, "ORA_results_exp1_combined.csv", row.names = FALSE)
```




```{r ORA_exp1, results='hide'}
# load in the gene lists for experiments 1 & 2;
exp1_DMR_list <- read.table("Exp1_DMRs_gene_id_matched.txt", header = T, sep = '\t')
exp1_DMR_list <- subset(exp1_DMR_list[exp1_DMR_list$p_fdr <= 0.1,])
exp1_DMR_list<- as.vector(exp1_DMR_list[,19])



# Run ORA
set.seed(100)
ORA_exp1_BP <- enricher(exp1_DMR_list, 
                TERM2GENE = GENES_BP,
                TERM2NAME = TERMs_BP,
                minGSSize=1,
                pvalueCutoff = 1,
                qvalueCutoff = 1)

set.seed(100)
ORA_exp1_MF <- enricher(exp1_DMR_list, 
                TERM2GENE = GENES_MF,
                TERM2NAME = TERMs_MF,
                minGSSize=1,
                pvalueCutoff = 1,
                qvalueCutoff = 1)

# Extract results and add in GO identifiers
ORA_exp1_BP_results <- ORA_exp1_BP@result
ORA_exp1_BP_results$GO <- "Biological Processes"

ORA_exp1_MF_results <- ORA_exp1_MF@result
ORA_exp1_MF_results$GO <- "Molecular Functions"

# combine results and subset for significant values
ORA_exp1_results <- rbind(ORA_exp1_BP_results, ORA_exp1_MF_results)
# ORA_exp1_results <- subset(ORA_exp1_results[ORA_exp1_results$p.adjust < 0.1,])

# Add column to id this as exp1;
ORA_exp1_results$experiment <- "Acute Experiment"

# Change fractions to numeric values
ORA_exp1_results$GeneRatio <- sapply(ORA_exp1_results$GeneRatio, convertFractionToNumeric)
ORA_exp1_results$BgRatio <- sapply(ORA_exp1_results$BgRatio, convertFractionToNumeric)

## Export Dataframe results
write.csv(ORA_exp1_results, "ORA_exp1_DMR_results.csv", row.names = FALSE)
```

## ORA results

```{r ORA_exp1_results, fig.height = 20, fig.width=15}
# Merged_ORA_exp1_BP$BgRatio <- sapply(merged_ORA_exp1_BP$BgRatio, convertFractionToNumeric)
ORA_exp1_results <- ORA_exp1_results[order(ORA_exp1_results$GeneRatio), ]


ORA_exp1_dotplot <- ggplot(data = ORA_exp1_results, 
                              aes(y = reorder(Description, 
                                              Count, 
                                              increasing = T), 
                                  x = GeneRatio, 
                                  color = p.adjust)) + 
                       geom_point(aes(size = Count), 
                                  stroke = 4) +
                       labs(title = "Biological Processes", y = "GO Description", x = "Gene Ratio") +
                       theme_light() + 
                       scale_color_gradient(low = "red", high = "blue") + 
                       scale_size_continuous(breaks = seq(min(ORA_exp1_results$Count), max(ORA_exp1_results$Count), by = 1)) +
                       theme(
                         legend.spacing = unit(0.5, "cm"),         # Adjust overall legend spacing
                         legend.key.height = unit(3, "lines"), # Adjust the height between size examples
                         axis.text.y = element_text(size = 11)
                         ) + 
                       guides(size = guide_legend(title = "Count"))  + 
                       scale_y_discrete(labels = function(x) str_wrap(x, width = 28)) + # Adjust width for line breaks 
                       facet_wrap(~GO, scales = "free")

# Merge ORA_exp1_BP results and dotplot together, and rm dotplot
plot(ORA_exp1_dotplot)

# Save current Environemnt:
save.image(file = "GSEA_exp1_prep.RData")
```

```{r GSEA_exp1, results = 'hide'}
# import dataframes
exp1_DM_region_data <- read.table("Exp1_DMRs_gene_id_matched.txt", header = T, sep = '\t')
exp1_DM_region_data$region <- as.factor(exp1_DM_region_data$region)
summary(exp1_DM_region_data$region)
# removing repeats to see if this make GSEA more accurate
  # exp1_DM_region_data <- subset(exp1_DM_region_data[exp1_DM_region_data$region != "dispersed_repeat",])
  # even with removing the repeats from this, there were still issue with a high degree of preranked ties.
  # This seems to be due to gene bodies potentially, I'm going to remove gene bodies to see how much this 
  # affects the ties percent

# exp1_DM_region_data <- exp1_DM_region_data[exp1_DM_region_data$region != "gene",]
# exp1_DM_region_data <- exp1_DM_region_data[exp1_DM_region_data$region != "dispersed_repeat",]
  # ties goes from 48% to 43%
  # with both repeats and genes removed, ties go down from 48% to 30%

# What about trying to perform the GSEAs separately for each region type?
#exp1_DM_region_data <- exp1_DM_region_data[exp1_DM_region_data$region == "gene",]
  # This works for gene, exon, promoter and repeats, and nothing else. I need a way to alter the ranking of the lists to make them better

# Combine gene ID and region into a single unique identifier
exp1_DM_region_data$gene_region_id <- paste0(exp1_DM_region_data[, 19], "_", exp1_DM_region_data$region)

# Generate GeneList Dataframe:
# Feature 1: numeric vector (meth.diff)
geneList_1 = exp1_DM_region_data[,7]

# Feature 2: Named Vector (rownames)
names(geneList_1) = as.character(exp1_DM_region_data[,20])

# add random jitter
set.seed(123)
geneList_1 <- geneList_1 + runif(length(geneList_1), min = -1e-6, max = 1e-6)
geneList_1 <- sort(geneList_1, decreasing = TRUE)

# Define region types to append
region_list <- c("gene", "exon", "five_prime_UTR", "three_prime_UTR", 
                 "dispersed_repeat", "promoter", "downstream_region", "ncRNA_gene")

# Create an empty list to hold expanded dataframes
expanded_list <- list()

# Loop through each region and duplicate rows
for (region in region_list) {
  temp_df <- GENES_BP  # Copy original TERM2GENE structure
  temp_df$GENE <- paste0(temp_df$GENE, "_", region)  # Append region to gene name
  expanded_list[[region]] <- temp_df  # Store in list
}

# Combine all expanded entries into a single dataframe
GENES_BP_expanded <- do.call(rbind, expanded_list)

for (region in region_list) {
  temp_df <- GENES_MF  # Copy original TERM2GENE structure
  temp_df$GENE <- paste0(temp_df$GENE, "_", region)  # Append region to gene name
  expanded_list[[region]] <- temp_df  # Store in list
}

# Combine all expanded entries into a single dataframe
GENES_MF_expanded <- do.call(rbind, expanded_list)

tail(geneList_1)
summary(geneList_1)


# Run GSEA exp1
set.seed(100)
GSEA_exp1_BP<- GSEA(geneList_1,
           TERM2GENE = GENES_BP_expanded,
           TERM2NAME = TERMs_BP,
           pvalueCutoff = 0.1)
set.seed(100)
GSEA_exp1_MF<- GSEA(geneList_1,
           TERM2GENE = GENES_MF_expanded,
           TERM2NAME = TERMs_MF,
           pvalueCutoff = 0.1)

# Get Results exp1
GSEA_exp1_BP_results <- GSEA_exp1_BP@result
GSEA_exp1_BP_results$GO <- "biological_processes"

GSEA_exp1_MF_results <- GSEA_exp1_MF@result
GSEA_exp1_MF_results$GO <- "molecular_function"

GSEA_exp1_results <- rbind(GSEA_exp1_BP_results, GSEA_exp1_MF_results)

# Subset for only significant results
GSEA_exp1_results <- subset(GSEA_exp1_results[GSEA_exp1_results$p.adjust < 0.1,])

# change function to work
# Make function to capture methylation data for each gene in the GSEA results
get_methDiff_data <- function(core_enrichment, meth_data) {
  genes <- unlist(strsplit(core_enrichment, "/"))
  lfc_values <- meth_data$meth.diff[match(genes, meth_data$gene_region_id)]
  paste(lfc_values, collapse = "/")
}

# Aplpy function to get methylation difference data
GSEA_exp1_results_with_methDiff <- GSEA_exp1_results %>%
  rowwise() %>%
  mutate(methDiff_combined = get_methDiff_data(core_enrichment, exp1_DM_region_data))

# Working on figures
GSEA_exp1_results_with_methDiff$Gene_IDs_list <- strsplit(as.character(GSEA_exp1_results_with_methDiff$core_enrichment), "/")
GSEA_exp1_results_with_methDiff$MethDiff_list <- strsplit(as.character(GSEA_exp1_results_with_methDiff$methDiff_combined), "/")

expanded_data_exp1 <- GSEA_exp1_results_with_methDiff %>% tidyr::unnest(c(Gene_IDs_list, MethDiff_list))
expanded_data_exp1$MethDiff_list <- as.numeric(expanded_data_exp1$MethDiff_list)

data_unique_exp1 <- expanded_data_exp1 %>% distinct()
GO_data_Exp1 <- data_unique_exp1

GO_data_Exp1$WrappedDescription <- str_wrap(GO_data_Exp1$Description, width = 20)  # Adjust the width as needed

GSEA_exp1_plot <- ggplot(GO_data_Exp1, aes(
  x = reorder(WrappedDescription, MethDiff_list), 
  y = MethDiff_list
)) +
  geom_half_boxplot(
    aes(fill = enrichmentScore,
        group = interaction(WrappedDescription, ID)),
    side = "l",   # left half
    alpha = 0.8
  ) +
  geom_half_point(
    aes(colour = enrichmentScore,
        group = interaction(WrappedDescription, ID)),
    side = "r",   # right half
    alpha = 0.4,
    size = 2
  ) +
  coord_flip() + 
  theme_classic()+
  facet_wrap(~GO, scales = "free" ,
             labeller = as_labeller(
               c(
                 "biological_processes" = "Biological Processes",
                 "molecular_function" = "Molecular Functions")
             )
  ) +
 scale_fill_gradient(low = "#2396de", high = "#db183f" , limits = c(-1,1))+
 scale_colour_gradient(low = "#2396de", high = "#db183f", limits = c(-1,1), 
                       guide = "none") +
  geom_hline(yintercept = 0,  linetype = "dashed",  color = "black", linewidth = 0.7) +
  labs(
    y = "Methylation Difference (%)", 
    x = "Gene Ontology (GO)", 
    fill = "Enrichment Score"
  )

print(GSEA_exp1_plot)
```

```{r looping_GSEA_by_region_exp1}
library(clusterProfiler)
library(dplyr)
library(ggplot2)
library(ggforce)
library(tidyr)
library(stringr)

# Read and prep data
exp1_DM_region_data <- read.table("Exp1_DMRs_gene_id_matched.txt", header = TRUE, sep = '\t')
exp1_DM_region_data$region <- as.factor(exp1_DM_region_data$region)

# Make df to store data
GSEA_exp1_data <- data.frame()

# temp region_type
# region_type = "gene"

# Loop over each region
for (region_type in levels(exp1_DM_region_data$region)) {
  
  message("Processing region: ", region_type)
  
  # Subset to region
  region_data <- exp1_DM_region_data[exp1_DM_region_data$region == region_type, ]
  
  # Build gene list
  geneList <- region_data[, 7]
  names(geneList) <- as.character(region_data[, 19])
  geneList <- sort(geneList, decreasing = TRUE)
  
  # Skip if too few genes
  if (length(geneList) < 50) {
    message("Skipping ", region_type, " — too few genes.")
    next
  }
  
  # Try GSEA for BP
  GSEA_BP <- tryCatch({
    set.seed(100)
    res <- GSEA(geneList, TERM2GENE = GENES_BP, TERM2NAME = TERMs_BP, pvalueCutoff = 0.1)
    res@result$GO <- "biological_processes"
    res@result
  }, error = function(e) {
    message("No BP enrichment for ", region_type)
    return(NULL)
  })
  
  # Try GSEA for MF
  GSEA_MF <- tryCatch({
    set.seed(100)
    res <- GSEA(geneList, TERM2GENE = GENES_MF, TERM2NAME = TERMs_MF, pvalueCutoff = 0.1)
    res@result$GO <- "molecular_function"
    res@result
  }, error = function(e) {
    message("No MF enrichment for ", region_type)
    return(NULL)
  })
  
  # Skip if both GSEAs failed
  if (is.null(GSEA_BP) && is.null(GSEA_MF)) {
    message("Skipping ", region_type, " — no significant enrichment in BP or MF.")
    next
  }
  
  # Combine and filter results
  all_res <- bind_rows(GSEA_BP, GSEA_MF)
  all_res$region <- region_type
  sig_res <- all_res[all_res$p.adjust < 0.1, ]
  
  if (nrow(sig_res) == 0) {
    message("No significant terms after p.adjust filtering for ", region_type)
    next
  }
  
  # Extract methylation differences per core enrichment
  sig_res <- sig_res %>%
    rowwise() %>%
    mutate(methDiff_combined = get_methDiff_data(core_enrichment, region_data))
  
  sig_res$Gene_IDs_list <- strsplit(as.character(sig_res$core_enrichment), "/")
  sig_res$MethDiff_list <- strsplit(as.character(sig_res$methDiff_combined), "/")
  
  expanded <- sig_res %>%
    tidyr::unnest(c(Gene_IDs_list, MethDiff_list))
  expanded$MethDiff_list <- as.numeric(expanded$MethDiff_list)
  expanded <- expanded %>% distinct()
  expanded$WrappedDescription <- str_wrap(expanded$Description, width = 20)
  # Plot
  g <- ggplot(expanded, aes(
    x = reorder(WrappedDescription, MethDiff_list),
    y = MethDiff_list
  )) +
    geom_half_boxplot(
      aes(fill = enrichmentScore,
          group = interaction(WrappedDescription, ID)),
      side = "l",
      alpha = 0.8
    ) +
    geom_half_point(
      aes(colour = enrichmentScore,
          group = interaction(WrappedDescription, ID)),
      side = "r",
      alpha = 0.4,
      size = 2
    ) +
    coord_flip() +
    theme_classic() +
    facet_wrap(~GO, scales = "free",
               labeller = as_labeller(
                 c("biological_processes" = "Biological Processes",
                   "molecular_function" = "Molecular Functions")
               )
    ) +
    scale_fill_gradient(low = "#2396de", high = "#db183f", limits = c(-1,1)) +
    scale_colour_gradient(low = "#2396de", high = "#db183f", limits = c(-1,1), guide = "none") +
    geom_hline(yintercept = 0, linetype = "dashed", color = "black", linewidth = 0.7) +
    labs(
      y = "Methylation Difference (%)",
      x = "Gene Ontology (GO)",
      fill = "Enrichment Score",
      title = paste("GSEA for region:", region_type)
    )
  
  # Save plot
  out_file <- paste0("GSEA_", region_type, "_plot.png")
  ggsave(out_file, g, width = 18, height = 12)
  
  message("Saved plot to ", out_file)
  
    # Add data to running file
  GSEA_exp1_data <- rbind(GSEA_exp1_data, sig_res)
  
}

# Import QUICKGO names
gsea_exp1_qgo_names <- read.table("GSEA_exp1_QUICKGO_names.tsv", header = F, sep = '\t')
gsea_exp1_qgo_names$Description2 <- gsea_exp1_qgo_names$V3
gsea_exp1_qgo_names$ID <- gsea_exp1_qgo_names$V1

# Merge dataframes to get new names
merged_gsea_exp1 <- merge(GSEA_exp1_data, gsea_exp1_qgo_names, by = "ID")

# Continue to plotting
{
  expanded_exp1 <- merged_gsea_exp1 %>%
    tidyr::unnest(c(Gene_IDs_list, MethDiff_list))
  expanded_exp1$MethDiff_list <- as.numeric(expanded_exp1$MethDiff_list)
  expanded_exp1 <- expanded_exp1 %>% distinct()
  expanded_exp1$WrappedDescription <- str_wrap(expanded_exp1$Description2, width = 20)
}


  # Plot
  g <- ggplot(expanded_exp1, aes(
    x = reorder(WrappedDescription, MethDiff_list),
    y = MethDiff_list
  )) +
    geom_half_boxplot(
      aes(group = interaction(WrappedDescription, ID)),
      fill = "lightgrey",
      side = "l",
      alpha = 1
    ) +
    geom_half_point(
      aes(colour = region),
      side = "r",
      alpha = 0.7,
      size = 3,
      position = position_nudge(x = 0)
    ) +
    coord_flip() +
    theme_classic() +
    facet_wrap(~GO, scales = "free",
               labeller = as_labeller(
                 c("biological_processes" = "Biological Processes",
                   "molecular_function" = "Molecular Functions")
               )
    ) +

    geom_hline(yintercept = 0, linetype = "dashed", color = "black", linewidth = 0.7) +
    labs(
      y = "Methylation Difference (%)",
      x = "Gene Ontology (GO)",
      fill = "Enrichment Score",
      title ="GSEA for Acute Experiment"
    ) +

  scale_color_manual(
    name = "Region",
    values = c("red", "#a1d61a","#0072B2", "#CC79A7"),
    labels = c("Repeat","Exon", "Gene", "Promoter")
  )
  
  g
  
```

# stratified ORA exp2

```{r stratified_ORA_exp2}
# Load required packages
library(clusterProfiler)

# Load DMR data
dmr_data <- read.table("exp2_DMRs_gene_id_matched.txt", header = TRUE, sep = '\t', stringsAsFactors = FALSE)

# Set the target region to analyze
  # promoter # results
  # three_prime_UTR
  # exon # results
  # gene # results
  # five_prime_UTR
  # downstream_region # results
  # dispersed_repeat  # results
  # ncRNA_gene
{
target_region <- "dispersed_repeat"  # Change this to the region you want to analyze

# Subset data to selected region
region_data <- dmr_data[dmr_data$region == target_region, ]

# Filter for significant DMRs
sig_data <- region_data[region_data$p_fdr <= 0.1, ]

# keep the gene-context mapping from the filtered DMR data:
gene_context_map <- sig_data[!is.na(sig_data$gene_id_labelling) & sig_data$gene_id_labelling != "", 
                             c("gene_id_labelling", "context")]


# Extract gene IDs, removing NA or blank
gene_ids <- as.character(sig_data$gene_id_labelling)
gene_ids <- gene_ids[!is.na(gene_ids) & gene_ids != ""]

# Proceed only if there are valid gene IDs
if (length(gene_ids) > 0) {

  # Run ORA for Biological Processes (BP)
  ORA_BP <- enricher(gene = gene_ids,
                     TERM2GENE = GENES_BP,
                     TERM2NAME = TERMs_BP,
                     minGSSize = 1,
                     pvalueCutoff = 1,
                     qvalueCutoff = 1)

  # Run ORA for Molecular Functions (MF)
  ORA_MF <- enricher(gene = gene_ids,
                     TERM2GENE = GENES_MF,
                     TERM2NAME = TERMs_MF,
                     minGSSize = 1,
                     pvalueCutoff = 1,
                     qvalueCutoff = 1)

  # Extract results
  bp_results <- ORA_BP@result
    # If geneID is delimited (e.g., "GENE1/GENE2/GENE3"), split it
  bp_expanded <- bp_results %>%
    mutate(geneID = strsplit(as.character(geneID), "/")) %>%
    unnest(geneID)
  # Add context by joining
  bp_expanded <- left_join(bp_expanded, gene_context_map,
                           by = c("geneID" = "gene_id_labelling"),
                           relationship = "many-to-many")
  

  mf_results <- ORA_MF@result
      # If geneID is delimited (e.g., "GENE1/GENE2/GENE3"), split it
  mf_expanded <- mf_results %>%
    mutate(geneID = strsplit(as.character(geneID), "/")) %>%
    unnest(geneID)
  # Add context by joining
  mf_expanded <- left_join(mf_expanded, gene_context_map,
                           by = c("geneID" = "gene_id_labelling"),
                           relationship = "many-to-many")

  # Tag result source
  bp_expanded$GO <- "Biological Processes"
  mf_expanded$GO <- "Molecular Functions"

  # Combine and annotate results
  all_results <- rbind(bp_expanded, mf_expanded)
  all_results$region <- target_region
  all_results$experiment <- "Primed Experiment"

  # Convert ratios to numeric
  convertFractionToNumeric <- function(x) {
    parts <- strsplit(as.character(x), "/")[[1]]
    if (length(parts) == 2) {
      return(as.numeric(parts[1]) / as.numeric(parts[2]))
    } else {
      return(NA)
    }
  }

  all_results$GeneRatio <- sapply(all_results$GeneRatio, convertFractionToNumeric)
  all_results$BgRatio <- sapply(all_results$BgRatio, convertFractionToNumeric)

  # Save to CSV
  assign(paste0("ORA_results_", target_region, "_exp2"), all_results)
  output_file <- paste0("ORA_results_", target_region, ".csv")
  write.csv(all_results, output_file, row.names = FALSE)

  message("Analysis complete and results saved.")
} else {
  message("No significant gene IDs found for the selected region.")
}
}

# Combined all results together
ORA_results_exp2_combined <- rbind(ORA_results_dispersed_repeat_exp2,
                                   ORA_results_downstream_region_exp2,
                                   ORA_results_exon_exp2,
                                   ORA_results_gene_exp2,
                                   ORA_results_promoter_exp2)

# Plotting Data
ORA_results_exp2_combined <- ORA_results_exp2_combined[order(ORA_results_exp2_combined$GeneRatio), ]

ORA_results_exp2_combined_clean <- unique(ORA_results_exp2_combined)


ORA_exp2_dotplot <- ggplot(data = ORA_results_exp2_combined_clean, 
                              aes(y = reorder(Description, 
                                              Count, 
                                              increasing = T), 
                                  x = GeneRatio, 
                                  color = p.adjust)) + 
                       geom_point(aes(size = Count), 
                                  stroke = 4) +
                       labs(title = "Biological Processes", y = "GO Description", x = "Gene Ratio") +
                       theme_light() + 
                       scale_color_gradient(low = "red", high = "blue") + 
                       scale_size_continuous(breaks = seq(min(ORA_results_exp2_combined$Count), max(ORA_results_exp2_combined$Count), by = 1)) +
                       theme(
                         legend.spacing = unit(0.5, "cm"),         # Adjust overall legend spacing
                         legend.key.height = unit(3, "lines"), # Adjust the height between size examples
                         axis.text.y = element_text(size = 11)
                         ) + 
                       guides(size = guide_legend(title = "Count"))  + 
                       scale_y_discrete(labels = function(x) str_wrap(x, width = 28)) + # Adjust width for line breaks 
                       facet_wrap(~GO, scales = "free")

# Merge ORA_exp2_BP results and dotplot together, and rm dotplot
plot(ORA_exp2_dotplot)

## Export Dataframe results
write.csv(ORA_results_exp2_combined_clean, "ORA_results_exp2_combined.csv", row.names = FALSE)


# Combined all results together
ORA_results_exp2_combined <- rbind(ORA_results_dispersed_repeat_exp2,
                                   ORA_results_exon_exp2,
                                   ORA_results_gene_exp2,
                                   ORA_results_downstream_region_exp2,
                                   ORA_results_promoter_exp2)

# Plotting Data
ORA_results_exp2_combined <- ORA_results_exp2_combined[order(ORA_results_exp2_combined$GeneRatio), ]


ORA_exp2_dotplot <- ggplot(data = ORA_results_exp2_combined, 
                              aes(y = reorder(Description, 
                                              Count, 
                                              increasing = T), 
                                  x = GeneRatio, 
                                  color = p.adjust)) + 
                       geom_point(aes(size = Count), 
                                  stroke = 4) +
                       labs(title = "Biological Processes", y = "GO Description", x = "Gene Ratio") +
                       theme_light() + 
                       scale_color_gradient(low = "red", high = "blue") + 
                       scale_size_continuous(breaks = seq(min(ORA_results_exp2_combined$Count), max(ORA_results_exp2_combined$Count), by = 1)) +
                       theme(
                         legend.spacing = unit(0.5, "cm"),         # Adjust overall legend spacing
                         legend.key.height = unit(3, "lines"), # Adjust the height between size examples
                         axis.text.y = element_text(size = 11)
                         ) + 
                       guides(size = guide_legend(title = "Count"))  + 
                       scale_y_discrete(labels = function(x) str_wrap(x, width = 28)) + # Adjust width for line breaks 
                       facet_wrap(~GO, scales = "free")

# Merge ORA_exp2_BP results and dotplot together, and rm dotplot
plot(ORA_exp2_dotplot)

## Export Dataframe results
write.csv(ORA_results_exp2_combined, "ORA_results_exp2_combined.csv", row.names = FALSE)

# Combining exp1 & exp2 data
ORA_results_combined <- rbind(ORA_results_exp1_combined_clean,
                              ORA_results_exp2_combined_clean)

ORA_results_combined$experiment <- factor(ORA_results_combined$experiment)
#ORA_results_combined <- ORA_results_combined[ORA_results_combined$p.adjust <= 0.1,]
ORA_results_combined <- ORA_results_combined[order(ORA_results_combined$GeneRatio), ]

# Lollipop plot
ora_lollipop <- ggplot(ORA_results_combined, 
       aes(y = reorder(Description, GeneRatio, FUN = min), 
           x = GeneRatio,
           color = region)) + 
  geom_segment(aes(x = 0, 
                   xend = GeneRatio, 
                   yend = Description), 
               color = "grey60", linewidth = 1) + 
  geom_point(aes(size = Count, color = region)) + 
  #scale_color_gradient(low = "#1675e0", high = "#e8c410", name = "p.adjust") +
    scale_color_manual(
    name = "Region",
    values = c("#E69F00", "#56B4E9", "#009E73",  "#0072B2",  "#CC79A7"),
    labels = c("Repeat", "Downstream", "Exon", "Gene", "Promoter")
  ) +
  scale_size_continuous(name = "Gene Count", range = c(4, 10)) +
  labs(x = "Gene Ratio",
       y = "GO Term") +
  theme_light() +
  theme(
    axis.text.y = element_text(size = 8),
    axis.text.x = element_text(size = 8),
    axis.title = element_text(size = 10),
    strip.text = element_text(size = 10),
    plot.title = element_text(hjust = 0.5, size = 14, face = "bold"),
    legend.title = element_text(size = 10),
    legend.text = element_text(size = 9)
  ) +
  scale_y_discrete(labels = function(x) str_wrap(x, width = 45)) +
  facet_wrap(~experiment~GO, scales = "free_y") +
  scale_x_continuous(limits = c(0, 1), breaks = c(0,0.2,0.4,0.6,0.8,1))

ggsave("ORA_lollipop_DMRs.png", ora_lollipop, width = 25, height = 25, units = "cm")


## Export Dataframe resultsexperiment## Export Dataframe results
write.csv(ORA_results_combined, "ORA_results_combined.csv", row.names = FALSE)

# read QUICKGO version of combined ORA
ORA_results_combined_named <- read.csv("ORA_results_combined_QUICKGO.csv", header = T, sep = ',')
ORA_results_combined_named$experiment <- factor(ORA_results_combined_named$experiment)
ORA_results_combined_named <- ORA_results_combined_named[order(ORA_results_combined_named$GeneRatio), ]
ORA_results_combined_named$enrichment <- ifelse(
  ORA_results_combined_named$p.adjust < 0.1,
  "enriched",
  "non-enriched"
)
ORA_results_combined_named$Count <- as.integer(ORA_results_combined_named$Count)  

# Lollipop plot
ORA_lolli <- ggplot(ORA_results_combined_named[ORA_results_combined_named$experiment == "Primed Experiment",], 
       aes(y = reorder(Description2, GeneRatio, FUN = min), 
           x = GeneRatio,
           color = region)) + 
  geom_segment(aes(x = 0, 
                   xend = GeneRatio, 
                   yend = Description2), 
               color = "grey60", linewidth = 1) + 
  geom_point(aes(size = Count, color = region)) + 
  #scale_color_gradient(low = "#1675e0", high = "#e8c410", name = "p.adjust") +
    scale_color_manual(
    name = "Region",
    values = c("#E69F00", "#56B4E9", "#009E73",  "#0072B2",  "#CC79A7"),
    labels = c("Repeat", "Downstream", "Exon", "Gene", "Promoter")
  ) +
  scale_size_continuous(
  name = "Gene Count", 
  range = c(2, 6),
  breaks = sort(unique(ORA_results_combined_named$Count)), 
  labels = function(x) as.integer(x)
) +
  theme_light() +
  theme(
    axis.text.y = element_text(size = 7),
    axis.text.x = element_text(size = 10),
    axis.title = element_text(size = 12),
    strip.text = element_text(size = 12),
    legend.title = element_text(size = 10),
    legend.text = element_text(size = 9),

  ) +
  scale_y_discrete(labels = function(x) str_wrap(x, width = 45)) +
  facet_wrap(~enrichment~GO, scales = "free_y") +
  scale_x_continuous(limits = c(0, 0.8), breaks = c(0,0.2,0.4,0.6,0.8)) +
  theme(
    strip.text = element_text(size = 10, margin = margin(0, 0, 2, 0)) #, strip.background = element_blank()
  ) +
  labs(
  x = "Gene Ratio",
  y = "Description"
)


ORA_lolli

ggsave("ORA_lollipop_plot_exp2.png", ORA_lolli, width = 25, height = 25, units = "cm")

```





```{r ORA_exp2, results='hide'}
# Run ORA
set.seed(100)
ORA_exp2_BP <- enricher(exp2_DMR_list, 
                TERM2GENE = GENES_BP,
                TERM2NAME = TERMs_BP,
                minGSSize=1,
                pvalueCutoff = 1,
                qvalueCutoff = 1)

set.seed(100)
ORA_exp2_MF <- enricher(exp2_DMR_list, 
                TERM2GENE = GENES_MF,
                TERM2NAME = TERMs_MF,
                minGSSize=1,
                pvalueCutoff = 1,
                qvalueCutoff = 1)

# Extract results and add in GO identifiers
ORA_exp2_BP_results <- ORA_exp2_BP@result
ORA_exp2_BP_results$GO <- "Biological Processes"

ORA_exp2_MF_results <- ORA_exp2_MF@result
ORA_exp2_MF_results$GO <- "Molecular Functions"

# combine results and subset for significant values
ORA_exp2_results <- rbind(ORA_exp2_BP_results, ORA_exp2_MF_results)
# ORA_exp2_results <- subset(ORA_exp2_results[ORA_exp2_results$p.adjust < 0.1,])

# Adding experiment column to differentiate between experiemnts for plotting
ORA_exp2_results$experiment <- "Primed Experiment"

# Change fractions to numeric values
ORA_exp2_results$GeneRatio <- sapply(ORA_exp2_results$GeneRatio, convertFractionToNumeric)
ORA_exp2_results$BgRatio <- sapply(ORA_exp2_results$BgRatio, convertFractionToNumeric)

## Export Dataframe results
write.csv(ORA_exp2_results, "ORA_exp2_DMR_results.csv", row.names = FALSE)
```

## ORA results

```{r ORA_exp2_results, fig.height = 8, fig.width=12}
# Merged_ORA_exp2_BP$BgRatio <- sapply(merged_ORA_exp2_BP$BgRatio, convertFractionToNumeric)
ORA_exp2_results <- ORA_exp2_results[order(ORA_exp2_results$GeneRatio), ]


ORA_exp2_dotplot <- ggplot(data = ORA_exp2_results, 
                              aes(y = reorder(Description, 
                                              Count, 
                                              increasing = T), 
                                  x = GeneRatio, 
                                  color = p.adjust)) + 
                       geom_point(aes(size = Count), 
                                  stroke = 4) +
                       labs(y = "GO Description", x = "Gene Ratio") +
                       theme_light() + 
                       scale_color_gradient(low = "red", high = "blue") + 
                       scale_size_continuous(breaks = seq(min(ORA_exp2_results$Count), max(ORA_exp2_results$Count), by = 1)) +
                       theme(
                         legend.spacing = unit(0.5, "cm"),         # Adjust overall legend spacing
                         legend.key.height = unit(3, "lines"), # Adjust the height between size examples
                         axis.text.y = element_text(size = 10)
                         ) + 
                       guides(size = guide_legend(title = "Count"))  + 
                       scale_y_discrete(labels = function(x) str_wrap(x, width = 20)) + # Adjust width for line breaks 
                       facet_wrap(~GO, scales = "free")

plot(ORA_exp2_dotplot)


# Save current Environemnt:
save.image(file = "GSEA_exp2_prep.RData")
```


## GSEA exp2


```{r GSEA_exp2, results = 'hide'}
# import dataframes
exp2_DM_region_data <- read.table("Exp2_DMRs_gene_id_matched.txt", header = T, sep = '\t')

# Generate GeneList Dataframe:
# Feature 1: numeric vector (meth.diff)
geneList_2 = exp2_DM_region_data[,7]

# Feature 2: Named Vector (rownames)
names(geneList_2) = as.character(exp2_DM_region_data[,19])

# feature 3: decreasing order;
geneList_2 = sort(geneList_2, decreasing = TRUE)

# Run GSEA exp2
set.seed(100)
GSEA_exp2_BP<- GSEA(geneList_2,
           TERM2GENE = GENES_BP,
           TERM2NAME = TERMs_BP,
           pvalueCutoff = 0.1)
set.seed(100)
GSEA_exp2_MF<- GSEA(geneList_2,
           TERM2GENE = GENES_MF,
           TERM2NAME = TERMs_MF,
           pvalueCutoff = 0.1)


# Get Results exp2
GSEA_exp2_BP_results <- GSEA_exp2_BP@result
GSEA_exp2_BP_results$GO <- "biological_processes"

GSEA_exp2_MF_results <- GSEA_exp2_MF@result
GSEA_exp2_MF_results$GO <- "molecular_function"

GSEA_exp2_results <- rbind(GSEA_exp2_BP_results, GSEA_exp2_MF_results)

GSEA_exp2_results <- subset(GSEA_exp2_results[GSEA_exp2_results$p.adjust < 0.1,])




GSEA_exp2_results_with_methDiff <- GSEA_exp2_results %>%
  rowwise() %>%
  mutate(methDiff_combined = get_methDiff_data(core_enrichment, exp2_DM_region_data))

# Working on figures
GSEA_exp2_results_with_methDiff$Gene_IDs_list <- strsplit(as.character(GSEA_exp2_results_with_methDiff$core_enrichment), "/")
GSEA_exp2_results_with_methDiff$MethDiff_list <- strsplit(as.character(GSEA_exp2_results_with_methDiff$methDiff_combined), "/")

expanded_data_exp2 <- GSEA_exp2_results_with_methDiff %>% tidyr::unnest(c(Gene_IDs_list, MethDiff_list))
expanded_data_exp2$MethDiff_list <- as.numeric(expanded_data_exp2$MethDiff_list)

data_unique_exp2 <- expanded_data_exp2 %>% distinct()
GO_data_Exp2 <- data_unique_exp2

GO_data_Exp2$WrappedDescription <- str_wrap(GO_data_Exp2$Description, width = 20)  # Adjust the width as needed

GSEA_exp2_plot <- ggplot(GO_data_Exp2, aes(
  x = reorder(WrappedDescription, MethDiff_list), 
  y = MethDiff_list
)) +
  geom_half_boxplot(
    aes(fill = enrichmentScore,
        group = interaction(WrappedDescription, ID)),
    side = "l",   # left half
    alpha = 0.8
  ) +
  geom_half_point(
    aes(colour = enrichmentScore,
        group = interaction(WrappedDescription, ID)),
    side = "r",   # right half
    alpha = 0.4,
    size = 2
  ) +
  coord_flip() + 
  theme_classic()+
  facet_wrap(~GO, scales = "free" ,
             labeller = as_labeller(
               c(
                 "biological_processes" = "Biological Processes",
                 "molecular_function" = "Molecular Functions")
             )
  ) +
 scale_fill_gradient(low = "#2396de", high = "#db183f" , limits = c(-1,1))+
 scale_colour_gradient(low = "#2396de", high = "#db183f", limits = c(-1,1), 
                       guide = "none") +
  geom_hline(yintercept = 0,  linetype = "dashed",  color = "black", linewidth = 0.7) +
  labs(
    y = "Methylation Difference (%)", 
    x = "Gene Ontology (GO)", 
    fill = "Enrichment Score"
  )

print(GSEA_exp2_plot)
```

# Loop for GSEA exp2

```{r looping_GSEA_by_region_exp2}
library(clusterProfiler)
library(dplyr)
library(ggplot2)
library(ggforce)
library(tidyr)
library(stringr)

# Read and prep data
exp2_DM_region_data <- read.table("exp2_DMRs_gene_id_matched.txt", header = TRUE, sep = '\t')
exp2_DM_region_data$region <- as.factor(exp2_DM_region_data$region)

# Make df to store data
GSEA_exp2_data <- data.frame()

# temp region_type
# region_type = "gene"

# Loop over each region
for (region_type in levels(exp2_DM_region_data$region)) {
  
  message("Processing region: ", region_type)
  
  # Subset to region
  region_data <- exp2_DM_region_data[exp2_DM_region_data$region == region_type, ]
  
  # Build gene list
  geneList <- region_data[, 7]
  names(geneList) <- as.character(region_data[, 19])
  geneList <- sort(geneList, decreasing = TRUE)
  
  # Skip if too few genes
  if (length(geneList) < 50) {
    message("Skipping ", region_type, " — too few genes.")
    next
  }
  
  # Try GSEA for BP
  GSEA_BP <- tryCatch({
    set.seed(100)
    res <- GSEA(geneList, TERM2GENE = GENES_BP, TERM2NAME = TERMs_BP, pvalueCutoff = 0.1)
    res@result$GO <- "biological_processes"
    res@result
  }, error = function(e) {
    message("No BP enrichment for ", region_type)
    return(NULL)
  })
  
  # Try GSEA for MF
  GSEA_MF <- tryCatch({
    set.seed(100)
    res <- GSEA(geneList, TERM2GENE = GENES_MF, TERM2NAME = TERMs_MF, pvalueCutoff = 0.1)
    res@result$GO <- "molecular_function"
    res@result
  }, error = function(e) {
    message("No MF enrichment for ", region_type)
    return(NULL)
  })
  
  # Skip if both GSEAs failed
  if (is.null(GSEA_BP) && is.null(GSEA_MF)) {
    message("Skipping ", region_type, " — no significant enrichment in BP or MF.")
    next
  }
  
  # Combine and filter results
  all_res <- bind_rows(GSEA_BP, GSEA_MF)
  all_res$region <- region_type
  sig_res <- all_res[all_res$p.adjust < 0.1, ]
  
  if (nrow(sig_res) == 0) {
    message("No significant terms after p.adjust filtering for ", region_type)
    next
  }
  
  # Extract methylation differences per core enrichment
  sig_res <- sig_res %>%
    rowwise() %>%
    mutate(methDiff_combined = get_methDiff_data(core_enrichment, region_data))
  
  sig_res$Gene_IDs_list <- strsplit(as.character(sig_res$core_enrichment), "/")
  sig_res$MethDiff_list <- strsplit(as.character(sig_res$methDiff_combined), "/")
  
  expanded <- sig_res %>%
    tidyr::unnest(c(Gene_IDs_list, MethDiff_list))
  expanded$MethDiff_list <- as.numeric(expanded$MethDiff_list)
  expanded <- expanded %>% distinct()
  expanded$WrappedDescription <- str_wrap(expanded$Description, width = 20)
  # Plot
  g <- ggplot(expanded, aes(
    x = reorder(WrappedDescription, MethDiff_list),
    y = MethDiff_list
  )) +
    geom_half_boxplot(
      aes(fill = enrichmentScore,
          group = interaction(WrappedDescription, ID)),
      side = "l",
      alpha = 0.8
    ) +
    geom_half_point(
      aes(colour = enrichmentScore,
          group = interaction(WrappedDescription, ID)),
      side = "r",
      alpha = 0.4,
      size = 2
    ) +
    coord_flip() +
    theme_classic() +
    facet_wrap(~GO, scales = "free",
               labeller = as_labeller(
                 c("biological_processes" = "Biological Processes",
                   "molecular_function" = "Molecular Functions")
               )
    ) +
    scale_fill_gradient(low = "#2396de", high = "#db183f", limits = c(-1,1)) +
    scale_colour_gradient(low = "#2396de", high = "#db183f", limits = c(-1,1), guide = "none") +
    geom_hline(yintercept = 0, linetype = "dashed", color = "black", linewidth = 0.7) +
    labs(
      y = "Methylation Difference (%)",
      x = "Gene Ontology (GO)",
      fill = "Enrichment Score",
      title = paste("GSEA for region:", region_type)
    )
  
  # Save plot
  out_file <- paste0("GSEA_", region_type, "_plot.png")
  ggsave(out_file, g, width = 18, height = 12)
  
  message("Saved plot to ", out_file)
  
    # Add data to running file
  GSEA_exp2_data <- rbind(GSEA_exp2_data, sig_res)
  
}

unique(GSEA_exp2_data$ID)

# Import QUICKGO names
gsea_exp2_qgo_names <- read.table("GSEA_exp2_QUICKGO_names.tsv", header = F, sep = '\t')
gsea_exp2_qgo_names$Description2 <- gsea_exp2_qgo_names$V3
gsea_exp2_qgo_names$ID <- gsea_exp2_qgo_names$V1

# Merge dataframes to get new names
merged_gsea_exp2 <- merge(GSEA_exp2_data, gsea_exp2_qgo_names, by = "ID")

# Continue to plotting
{
  expanded_exp2 <- merged_gsea_exp2 %>%
    tidyr::unnest(c(Gene_IDs_list, MethDiff_list))
  expanded_exp2$MethDiff_list <- as.numeric(expanded_exp2$MethDiff_list)
  expanded_exp2 <- expanded_exp2 %>% distinct()
  expanded_exp2$WrappedDescription <- str_wrap(expanded_exp2$Description2, width = 20)
}


  # Plot
  g <- ggplot(expanded_exp2, aes(
    x = reorder(WrappedDescription, MethDiff_list),
    y = MethDiff_list
  )) +
    geom_half_boxplot(
      aes(group = interaction(WrappedDescription, ID)),
      fill = "lightgrey",
      side = "l",
      alpha = 1
    ) +
    geom_half_point(
      aes(colour = region),
      side = "r",
      alpha = 0.7,
      size = 3,
      position = position_nudge(x = 0)
    ) +
    coord_flip() +
    theme_classic() +
    facet_wrap(~GO, scales = "free",
               labeller = as_labeller(
                 c("biological_processes" = "Biological Processes",
                   "molecular_function" = "Molecular Functions")
               )
    ) +

    geom_hline(yintercept = 0, linetype = "dashed", color = "black", linewidth = 0.7) +
    labs(
      y = "Methylation Difference (%)",
      x = "Gene Ontology (GO)",
      fill = "Enrichment Score"
    ) +

  scale_color_manual(
    name = "Region",
    values = c("red", "#a1d61a","#0072B2", "#CC79A7"),
    labels = c("Repeat","Exon", "Gene", "Promoter")
  )
  
  g
  
ggsave("GSEA_DMRs_exp2.png", g, width = 25, height = 12.5, units = "cm")
  
```

## Plot GSEAs together

```{r GSEAs_combined}
# combined data
GSEA_exp1_data$exp <- "exp1"
GSEA_exp1_data$setSize <- as.numeric(GSEA_exp1_data$setSize)
str(GSEA_exp1_data)
GSEA_exp2_data$exp <- "exp2"
GSEA_exp2_data$setSize <- as.numeric(GSEA_exp2_data$setSize)
GSEA_combined_data <- rbind(GSEA_exp1_data, GSEA_exp2_data)
GSEA_combined_data$exp <- as.factor(GSEA_combined_data$exp)

{
  expanded <- GSEA_combined_data %>%
    tidyr::unnest(c(Gene_IDs_list, MethDiff_list))
  expanded$MethDiff_list <- as.numeric(expanded$MethDiff_list)
  expanded <- expanded %>% distinct()
  expanded$WrappedDescription <- str_wrap(expanded$Description, width = 20)
}

 # Plot
  g <- ggplot(expanded, aes(
    x = reorder(WrappedDescription, MethDiff_list),
    y = MethDiff_list
  )) +
    geom_half_boxplot(
      aes(group = interaction(WrappedDescription, ID)),
      fill = "lightgrey",
      side = "l",
      alpha = 1
    ) +
    geom_half_point(
      aes(colour = region),
      side = "r",
      alpha = 0.5,
      size = 2.5,
      position = position_nudge(x = 0)
    ) +
    coord_flip() +
    theme_classic() +
    facet_wrap(~exp~GO, scales = "free",
               labeller = as_labeller(
                 c("biological_processes" = "Biological Processes",
                   "molecular_function" = "Molecular Functions",
                   "exp1" = "Acute",
                   "exp2" = "Primed")
               )
    ) +

    geom_hline(yintercept = 0, linetype = "dashed", color = "black", linewidth = 0.7) +
    labs(
      y = "Methylation Difference (%)",
      x = "Gene Ontology (GO)",
      fill = "Enrichment Score"
    ) +

  scale_color_manual(
    name = "Region",
    values = c("red", "#a1d61a","#0072B2", "#CC79A7"),
    labels = c("Repeat","Exon", "Gene", "Promoter")
  )
  
  g
  
ggsave("GSEA_DMRs.png", g, width = 25, height = 25, units = "cm")






# Now with QUICKGO names
  # combined expanded lists

expanded_combined <- rbind(expanded_exp1, expanded_exp2)

# plot
# Plot
  g <- ggplot(expanded_combined, aes(
    x = reorder(WrappedDescription, MethDiff_list),
    y = MethDiff_list
  )) +
    geom_half_boxplot(
      aes(group = interaction(WrappedDescription, ID)),
      fill = "lightgrey",
      side = "l",
      alpha = 1
    ) +
    geom_half_point(
      aes(colour = region),
      side = "r",
      alpha = 0.5,
      size = 2.5,
      position = position_nudge(x = 0)
    ) +
    coord_flip() +
    theme_classic() +
    facet_wrap(~exp~GO, scales = "free",
               labeller = as_labeller(
                 c("biological_processes" = "Biological Processes",
                   "molecular_function" = "Molecular Functions",
                   "exp1" = "Acute",
                   "exp2" = "Primed")
               )
    ) +

    geom_hline(yintercept = 0, linetype = "dashed", color = "black", linewidth = 0.7) +
    labs(
      y = "Methylation Difference (%)",
      x = "Gene Ontology (GO)",
      fill = "Enrichment Score"
    ) +

  scale_color_manual(
    name = "Region",
    values = c("red", "#a1d61a","#0072B2", "#CC79A7"),
    labels = c("Repeat","Exon", "Gene", "Promoter")
  )
  
  g

  ggsave("GSEA_DMRs.png", g, width = 25, height = 25, units = "cm")

  
```

## making table of GSEA results:

```{r GSEA_table}
#remove unnecessary columns
GSEA_results_pruned <- subset(GSEA_combined_data[c(1,2,3,5,7,11,12,13,17,14)],)
# Export GSEA results to get QUICKGO descriptions
write.csv(GSEA_results_pruned, "GSEA_combined_data.csv", row.names = FALSE)
#read in new data
GSEA_results_pruned <- read.csv("GSEA_combined_data_named.csv", header = T, sep = ",")

# rearrange to liking
GSEA_results_organised <- subset(GSEA_results_pruned[c(9,7,8,11,3,4,5)],)

# Re-order by experiment first, then GO within experiment
GSEA_results_organised <- GSEA_results_organised %>%
  arrange(exp, GO, region, Description2)

GSEA_results_organised <- GSEA_results_organised %>%
  mutate(
    exp = case_when(
      exp == "exp1" ~ "Acute",
      exp == "exp2" ~ "Primed",
      TRUE ~ exp
    ),
    GO = case_when(
      GO == "biological_processes" ~ "Biological Process",
      GO == "molecular_function" ~ "Molecular Function",
      TRUE ~ GO
    ),
    region = case_when(
      region == "dispersed_repeat" ~ "Repeats",
      region == "exon" ~ "Exon",
      region == "gene" ~ "Gene Body",
      region == "promoter" ~ "Promoter",
      TRUE ~ region
    )
  )

# Making table
# Find where the experiment changes (e.g., row index before second experiment starts)
exp_change_row <- 16

# Get row numbers where each region group ends
region_breaks <- GSEA_results_organised %>%
  mutate(row = row_number()) %>%
  group_by(exp, GO, region) %>%
  summarise(last_row = max(row), .groups = "drop") %>%
  pull(last_row)


set_flextable_defaults(background.color = "transparent")

ft <- flextable(GSEA_results_organised) %>%
  set_header_labels(
    exp = "Experiment", 
    GO = "Gene Ontology",
    region = "Region",
    Description2 = "Description",
    setSize = "Set Size",
    NES = "NES",
    p.adjust = "P-adj"
  ) %>%
  theme_booktabs() %>%

  # Slim down the column widths
  width(j = "exp", width = 0.5) %>%
  width(j = "GO", width = 1.2) %>%
  width(j = "region", width = 1.2) %>%
  width(j = "Description2", width = 1.5) %>%
  width(j = "setSize", width = 0.4) %>%
  width(j = "NES", width = 0.4) %>%
  width(j = "p.adjust", width = 0.5) %>%

  align(j = "Description2", align = "left", part = "all") %>%
  align(j = c("setSize", "NES", "p.adjust"), align = "center", part = "all") %>%

  merge_v(j = "exp") %>%
  valign(j = "exp", valign = "top") %>%
  merge_v(j = "region") %>%
  valign(j = "region", valign = "top") %>%
  merge_v(j = "GO") %>%
  valign(j = "GO", valign = "top") %>%

  hline(i = exp_change_row, border = fp_border(color = "black", width = 1)) %>%
  border(i = 1, border.top = fp_border(color = "black", width = 1.5)) %>%
  border(i = nrow(GSEA_results_organised), border.bottom = fp_border(color = "black", width = 1.5)) %>%

  bg(i = seq(2, nrow(GSEA_results_organised), by = 2), bg = "#F7F7F7", part = "body") %>%

  fontsize(part = "header", size = 8) %>%
  fontsize(part = "body", size = 7) %>%
  padding(part = "all", padding.top = 2, padding.bottom = 2, padding.left = 1, padding.right = 1) %>%

  colformat_double(j = "NES", digits = 3) %>%
  colformat_double(j = "p.adjust", digits = 3) %>%

  set_caption(caption = "Table 3. GSEA results of the Acute and Primed responses to Diesel exposure.")


ft

doc <- read_docx()
doc <- body_add_flextable(doc, value = ft, align = "left")
print(doc, target = "GSEA_table.docx")

```

# Calculating and comparing proportions of genome aganist DMR features:

```{r}

exp1_DMRs <- read.csv("exp1_DMRs_refined.csv", header = T)
exp1_DMRs <- subset(exp1_DMRs[exp1_DMRs$p_fdr <= 0.1,])

exp1_DMRs$region[exp1_DMRs$region == "three_prime_UTR"] <- "3` UTR"


exp2_DMRs <- read.table("exp2_DMRs.txt", header = T, sep = "\t")
exp2_DMRs <- subset(exp2_DMRs[exp2_DMRs$p_fdr <= 0.1,])

# If your columns are named differently, adjust these selections accordingly
df_summary_exp1 <- exp1_DMRs %>%
  group_by(region) %>%                # Adjust "feature" if your column has a different name
  summarize(feature_count = n()) %>%   # Count how many rows per feature
  mutate(proportion_percent = 100 * feature_count / sum(feature_count))

# If your columns are named differently, adjust these selections accordingly
df_summary_exp2 <- exp2_DMRs %>%
  group_by(region) %>%                # Adjust "feature" if your column has a different name
  summarize(feature_count = n()) %>%   # Count how many rows per feature
  mutate(proportion_percent = 100 * feature_count / sum(feature_count))

# View the summarized data
print(df_summary_exp1)
print(df_summary_exp2)

# Construct the summary of the combined_annotations.gff3 file:
combined_annotations_summary <- NULL

combined_annotations_summary$region <- c("Repeat",
                                         "Downstream",
                                         "exon",
                                         "5` UTR",
                                         "Gene",
                                          "ncRNA",
                                          "Promoter",
                                          "3` UTR")

combined_annotations_summary$feature_count <- c(978221,
                                                47671,
                                                383432,
                                                23401,
                                                47671,
                                                7980,
                                                47671,
                                                23356
                                                )



combined_annotations_summary$proportion_percent <- c(62.73,
                                                     3.06,
                                                     24.59,
                                                     1.50,
                                                     3.06,
                                                     0.51,
                                                     3.06,
                                                     1.50
                                                     )

combined_annotations_summary <- as.data.frame(combined_annotations_summary)

library(ggplot2)

# Sort from largest to smallest so that
# the largest portion is drawn first (on the bottom of the bar).
# df_summary <- df_summary %>%
#   arrange(proportion_percent)
# 
# # Turn 'region' into a factor with levels in descending order
# df_summary$region <- factor(df_summary$region, levels = df_summary$region)
# 
# # Bar plot by proportion
# ggplot(df_summary, aes(x = region, y = proportion_percent, fill = region)) +
#   geom_bar(stat = "identity") +
#   labs(title = "Bar Plot of Feature Proportions",
#        x = "Feature",
#        y = "Proportion (%)") +
#   theme_bw()
# 
# 
# # Sort from largest to smallest so that
# # the largest portion is drawn first (on the bottom of the bar).
# combined_annotations_summary <- combined_annotations_summary %>%
#   arrange(proportion_percent)
# 
# # Turn 'region' into a factor with levels in descending order
# combined_annotations_summary$region <- factor(combined_annotations_summary$region, levels = combined_annotations_summary$region)
# 
# # Bar plot by proportion
# ggplot(combined_annotations_summary, aes(x = region, y = proportion_percent, fill = region)) +
#   geom_bar(stat = "identity") +
#   labs(
#     title = "Bar Plot of Feature Proportions",
#     x = "Feature",
#     y = "Proportion (%)"
#   ) +
#   theme_bw() +
#   theme(
#     axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)
#   )



df_summary_exp1$type <- "DMRs - Acute"
df_summary_exp2$type <- "DMRs - Primed"
combined_annotations_summary$type <- "Annotated Genome"

all_data <- rbind(df_summary_exp1, df_summary_exp2, combined_annotations_summary)

all_data <- all_data %>%
  arrange(proportion_percent)

all_data$region <- as.factor(all_data$region)

# rename values to be correct
all_data$region[all_data$region == "dispersed_repeat"] <- "Repeat"
all_data$region[all_data$region == "promoter"] <- "Promoter"

# Set the order of feature_type as a factor
all_data$region <- factor(all_data$region, levels = c("Repeat", 
                                                                  "exon",
                                                                  "Promoter",
                                                                  "Gene",
                                                                  "Downstream",
                                                                  "3` UTR",
                                                                  "5` UTR",
                                                                  "ncRNA"))



rel_plot <- ggplot(all_data, aes(x = region, y = proportion_percent, fill = region)) +
  geom_bar(stat = "identity") +
  geom_text(aes(label = feature_count), vjust = -0.5, color = "black", angle = 45, hjust = 0)+
  labs(
    x = "Feature",
    y = "Proportion (%)"
  ) +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)
  ) + 
  facet_wrap(~type, nrow = 1, scales = "free_x") +
    scale_fill_manual(
    name = "Region",
    values = c("#E69F00", "#56B4E9", "#009E73", "#a1d61a", "#0072B2", "#D55E00", "#CC79A7", "red"),
    labels = c("Repeat", "Downstream", "Exon", "5' UTR", "Gene","ncRNA", "Promoter", "3' UTR")) 

rel_plot

ggsave("./rel_plot.png", rel_plot, width = 30, height = 20, units= "cm", dpi = 600)


abs_plot <- ggplot(all_data, aes(x = region, y = feature_count, fill = region)) +
  geom_bar(stat = "identity") +
  labs(
    x = "Feature",
    y = "Count of Features"
  ) +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)
  ) + 
  facet_wrap(~type, nrow = 1, scales = "free")



grid.arrange(abs_plot, rel_plot, nrow = 2)

install.packages("gtable")
library(gtable)

g1 <- ggplotGrob(abs_plot)
g2 <- ggplotGrob(rel_plot)
g <- rbind(g1, g2, size = "first")
g$widths <- unit.pmax(g1$widths, g2$widths)
grid.newpage()
grid.draw(g)


```

# Making a new figure with the uniprot matched GO ids and descriptions of DMRs

```{r}
data <- read.table("DMR functional terms table.txt", header = T, sep = "\t")

data <- data[ !apply(data =="", 1, any),]

data <- data %>%
  arrange(Methylation.Difference....)

data$Methylation.Difference.... <- as.numeric(data$Methylation.Difference....)
data$GO <- as.factor(data$GO)
data$GO.Type <- as.factor(data$GO.Type)
data$Feature.type <- as.factor(data$Feature.type)
data$Experiment <- as.factor(data$Experiment)
data$Cytosine.Context <- as.factor(data$Cytosine.Context)
data$Enriched <- as.factor(data$Enriched)

data$Wrapped_GO <- str_wrap(data$GO, width = 30)  # Adjust the width as needed


cols <- c("#e08282", "#f5e9e9")
cols2 <- c("red2", "#4d4949")

ggplot(
  data = data,
  aes(
    x = Methylation.Difference....,
    y = reorder(Wrapped_GO, Methylation.Difference....),
    color = Enriched,
    fill = Enriched
  )
) +
  geom_boxplot(alpha = 0.5) +
  facet_wrap(~Experiment + GO.Type, scales = "free") +
  labs(
    x = "Methylation Difference (%)",
    y = "Gene Ontology (GO)",
    fill = "Gene Ontology",
    color = "Gene Ontology",
    size = "Adjusted P-Value"
  ) +
  theme_bw() +
  scale_fill_manual(values = cols) +
  scale_colour_manual(values = cols2)

ggsave("ORA_DMRs.png",width = 25, height = 30, dpi = 300, units = "cm")

# Need to change the descriptions of the functionlly enriched terms so that I can highlight them on plot;
ORA_results_combined

write.table(ORA_results_combined, "ORA_results_combined_TO_BE_RENAMED.txt", sep = '\t')

ORA_results_combined_RENAMED <- read.table("ORA_results_combined_RENAMED.txt", sep = '\t', header = T)



# -----------------------------------------------------------------------------

# Making a plot of just the acute exposure expeirment
# including the ggside splits to show the number of parts associated to each function

data_acute <- data[data$Experiment == "Acute",]
data_acute$Feature.type <- as.character(data_acute$Feature.type)
str(data_acute)
data_primed <- data[data$Experiment == "Primed",]

# Change colour parameters here:
cols <- c("#e08282", "#f5e9e9")
cols2 <- c("red2", "#4d4949")

library(ggside)
library(gghalves)

# Acute plot
acute_fe_plot <- ggplot(
  data = data_acute,
  aes(
    x = Methylation.Difference....,
    y = reorder(Wrapped_GO, Methylation.Difference....),
    color = Enriched,
    fill = Enriched
  )
) +
  geom_boxplot(alpha = 0.5) +
  facet_wrap(~GO.Type, scales = "free",
             labeller = as_labeller(
               c(
                 "BP" = "Biological Processes",
                 "MF" = "Molecular Functions")
                )
             ) +
  labs(
    x = "Methylation Difference (%)",
    y = "Gene Ontology (GO)",
    fill = "Overrepresentation",
    color = "Overrepresentation",
    size = "Adjusted P-Value"
  ) +
  theme_bw() +
  scale_fill_manual(values = cols) +
  scale_colour_manual(values = cols2)

acute_fe_plot

ggsave("./acute_fe_plot.png", acute_fe_plot, width = 2, height = 16, units= "cm", dpi = 600)

# Primed plot
primed_fe_plot <- ggplot(
  data = data_primed,
  aes(
    x = Methylation.Difference....,
    y = reorder(Wrapped_GO, Methylation.Difference....),
    color = Enriched,
    fill = Enriched
  )
) +
  geom_boxplot(alpha = 0.5) +
  facet_wrap(~GO.Type, 
             scales = "free",
             labeller = as_labeller(
               c(
                 "BP" = "Biological Processes",
                 "MF" = "Molecular Functions")
                )
             ) +
  labs(
    x = "Methylation Difference (%)",
    y = "Gene Ontology (GO)",
    fill = "Overrepresentation",
    color = "Overrepresentation",
    size = "Adjusted P-Value"
  ) +
  theme_bw() +
  scale_fill_manual(values = cols) +
  scale_colour_manual(values = cols2)

primed_fe_plot

ggsave("./primed_fe_plot.png", primed_fe_plot, width = 24, height = 16, units= "cm", dpi = 600)

getwd()
```

# Correlating Expression and Methylation

```{r}

# Importing data file which already has expression and methylation combined from using the 'VLOOKUP' function in excel, I used the full DMR results for both experiments, alongside the full DEA results for both experiments, homogenised the naming conventions (removing .TU, and replacing .path with .mrna).


getwd()

meth_exp_data <- read.table("Corelating Methylation and Expression All context data.txt", header = T, sep = '\t')

str(meth_exp_data)

meth_exp_data <- meth_exp_data %>%
  mutate(region = ifelse(region == "dispersed_repeat", "Repeat", region))
meth_exp_data <- meth_exp_data %>%
  mutate(region = ifelse(region == "downstream_region", "Downstream", region))
meth_exp_data <- meth_exp_data %>%
  mutate(region = ifelse(region == "exon", "Exon", region))        
meth_exp_data <- meth_exp_data %>%
  mutate(region = ifelse(region == "five_prime_UTR", "5` UTR", region))
meth_exp_data <- meth_exp_data %>%
  mutate(region = ifelse(region == "ncRNA_gene", "ncRNA", region))
meth_exp_data <- meth_exp_data %>%
  mutate(region = ifelse(region == "promoter", "Promoter", region))
meth_exp_data <- meth_exp_data %>%
  mutate(region = ifelse(region == "three_prime_UTR", "3` UTR", region))
meth_exp_data <- meth_exp_data %>%
  mutate(region = ifelse(region == "gene", "Gene", region))
meth_exp_data <- meth_exp_data %>%
  mutate(experiment = ifelse(experiment == "exp1", "Acute Experiment", experiment))
meth_exp_data <- meth_exp_data %>%
  mutate(experiment = ifelse(experiment == "exp2", "Primed Experiment", experiment))


meth_exp_data$experiment <- as.factor(meth_exp_data$experiment)
meth_exp_data$context <- as.factor(meth_exp_data$context)
meth_exp_data$region <- as.factor(meth_exp_data$region)


library(ggplot2)

ggplot(meth_exp_data, 
       aes(x = meth.diff,
           y = exp_l2fc,
           col = region
           )) + 
  geom_point(alpha = 0.5, size = 3) + 
  geom_smooth(method = "lm", se = F, color = "black", lwd = 2) +
  geom_vline(xintercept = c(0), col = "black", linetype = "dashed") +
  geom_hline(yintercept = c(0), col = "black", linetype = "dashed") +
  scale_color_manual(
    name = "Region",
    values = c("#E69F00", "#56B4E9", "#009E73", "#a1d61a", "#0072B2", "#D55E00", "#CC79A7", "red"),
    labels = c("Repeat", "Downstream", "Exon", "5' UTR", "Gene","ncRNA", "Promoter", "3' UTR")) +
  labs(
    x = "Methylation Difference (%)",
    y = "Expression Change (Log2FoldChange)",
    color = "Region Type"
  ) +
  theme_bw() + 
  scale_x_continuous(limits = c(-40, 40), breaks = c(-40,-20,0,20,40)) + 
  scale_y_continuous(limits = c(-5, 5), breaks = c(-5,-4,-3, -2, -1, 0, 1, 2, 3,4,5)) + 
  facet_grid(~experiment~region)

# -----------------------------------------------------------------------------

# Making a nice set of plots

levels(meth_exp_data$region)
# reorder regions
meth_exp_data$region <- factor(
  meth_exp_data$region,
  levels = c("Repeat", "Downstream", "Exon", "5` UTR",
             "Gene", "ncRNA", "Promoter", "3` UTR")  # <- Your desired order
)

exp1_corr_plot <- ggplot(meth_exp_data[(meth_exp_data$experiment == "Acute Experiment") #& (meth_exp_data$meth_p_fdr <= 0.1)
                                       ,], 
       aes(x = meth.diff,
           y = exp_l2fc,
           col = region
           )) + 
  geom_point(alpha = 0.5, size = 2, aes(shape = context)) + 
  geom_smooth(method = "lm", se = F, color = "black", lwd = 1) +
  geom_vline(xintercept = c(0), col = "black", linetype = "dashed") +
  geom_hline(yintercept = c(0), col = "black", linetype = "dashed") +
  scale_color_manual(
    name = "Region",
    values = c("#E69F00", "#56B4E9", "#009E73", "#a1d61a", "#0072B2", "#D55E00", "#CC79A7", "red"),
    labels = c("Repeat", "Downstream", "Exon", "5' UTR", "Gene","ncRNA", "Promoter", "3' UTR")) +
  labs(
    x = "Methylation Difference (%)",
    y = "Expression Change (Log2FoldChange)",
    color = "Region Type"
  ) +
  theme_bw() + 
  #scale_x_continuous(limits = c(-40, 40), breaks = c(-40,-20,0,20,40)) + 
  scale_y_continuous(limits = c(-5, 5), breaks = c(-5,-4,-3, -2, -1, 0, 1, 2, 3,4,5)) + 
  facet_wrap(~region, nrow = 2, scales = "free")  + 
  guides(color = "none")

print(exp1_corr_plot)

ggsave("exp1_meth-expression_plot.png", exp1_corr_plot, width = 18, height = 10, units = "cm")

# Create an empty results dataframe
corr_results_exp1 <- data.frame(
  region = character(),
  correlation = numeric(),
  p_value = numeric(),
  stringsAsFactors = FALSE
)

meth_exp_data_exp1 <- meth_exp_data[meth_exp_data$experiment == "Acute Experiment",]

# Loop through each level of "region"
for (reg in levels(meth_exp_data_exp1$region)) {
  
  # Subset the data for this region
  region_data <- subset(meth_exp_data_exp1, region == reg)
  
  # Perform correlation test
  test <- cor.test(region_data$meth.diff, region_data$exp_l2fc)
  
  # Add the result to the results dataframe
  corr_results_exp1 <- rbind(corr_results_exp1, data.frame(
    region = reg,
    correlation = test$estimate,
    p_value = test$p.value
  ))
}


exp2_corr_plot <- ggplot(meth_exp_data[(meth_exp_data$experiment == "Primed Experiment") # & (meth_exp_data$meth_p_fdr <= 0.1) # also an option to look at only the significantly differentiall expressed regions.
                                       ,], 
       aes(x = meth.diff,
           y = exp_l2fc,
           col = region
           )) + 
  geom_point(alpha = 0.5, size = 2, aes(shape = context)) + 
  geom_smooth(method = "lm", se = F, color = "black", lwd = 1) +
  geom_vline(xintercept = c(0), col = "black", linetype = "dashed") +
  geom_hline(yintercept = c(0), col = "black", linetype = "dashed") +
  scale_color_manual(
    name = "Region",
    values = c("#E69F00", "#56B4E9", "#009E73", "#a1d61a", "#0072B2", "#D55E00", "#CC79A7", "red"),
    labels = c("Repeat", "Downstream", "Exon", "5' UTR", "Gene","ncRNA", "Promoter", "3' UTR")) +
  labs(
    x = "Methylation Difference (%)",
    y = "Expression Change (Log2FoldChange)",
    color = "Region Type"
  ) +
  theme_bw() + 
  #scale_x_continuous(limits = c(-40, 40), breaks = c(-40,-20,0,20,40)) + 
  scale_y_continuous(limits = c(-5, 5), breaks = c(-5,-4,-3, -2, -1, 0, 1, 2, 3,4,5)) + 
  facet_wrap(~region, nrow = 2, scales = "free")  + 
  guides(color = "none")

print(exp2_corr_plot)

ggsave("exp2_meth-expression_plot.png", exp2_corr_plot, width = 18, height = 10, units = "cm")


# Create an empty results dataframe
corr_results_exp2 <- data.frame(
  region = character(),
  correlation = numeric(),
  p_value = numeric(),
  stringsAsFactors = FALSE
)

meth_exp_data_exp2<- meth_exp_data[meth_exp_data$experiment == "Primed Experiment",]

corr_results_exp2 <- NULL

# Loop through each level of "region"
for (reg in levels(meth_exp_data_exp2$region)) {
  
  # Subset the data for this region
  region_data <- subset(meth_exp_data_exp2, region == reg)
  
  # Perform correlation test
  test <- cor.test(region_data$meth.diff, region_data$exp_l2fc)
  
  # Add the result to the results dataframe
  corr_results_exp2 <- rbind(corr_results_exp2, data.frame(
    region = reg,
    correlation = test$estimate,
    p_value = test$p.value
  ))
}


# -------------------------------------------------------------------------------

exp1_meth_exp_data <- meth_exp_data[meth_exp_data$experiment == "Acute Experiment",]
exp1_corr_results <- cor.test(exp1_meth_exp_data$meth.diff, 
                              exp1_meth_exp_data$exp.l2fc, 
                              method = "pearson")

print(exp1_corr_results)
  # p = 0.003123
  # correlation = -0.03447973 

summary(exp1_meth_exp_data)

exp2_meth_exp_data <- meth_exp_data[meth_exp_data$experiment == "Primed Experiment",]
exp2_corr_results <- cor.test(exp2_meth_exp_data$meth.diff, 
                              exp2_meth_exp_data$l2fc, 
                              method = "pearson")

print(exp2_corr_results)
  # p = 0.7574
  # correlation = -0.003096255 

summary(exp2_meth_exp_data)


lm <- lm(data = meth_exp_data,
         l2fc~meth.diff)
summary(lm)
anova(lm)

lm_exp1 <- lm(data = meth_exp_data[meth_exp_data$experiment == "Acute Experiment",],
         l2fc~meth.diff+region)
summary(lm_exp1)
anova(lm_exp1)

lm_exp2 <- lm(data = meth_exp_data[meth_exp_data$experiment == "Primed Experiment",],
         l2fc~meth.diff+region)
summary(lm_exp2)
anova(lm_exp2)


```



```{r}


```