---
title: "Differential Methylated Regions Functional Enrichment Analyses"
author: "Hamish Williams"
date: "`r Sys.Date()`"
output: word_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Overview

Performing over representation analysis on the DMR gene lists for both experiments to determine if there is some statistcially enriched terms and functions. This analysis is only going to contribute to highlighting if particular functions are being enriched, and not to remove genes associated to DMRs away from the results and discussion. The results of the DMR anlaysis have already gone through a lengthy and rigouros bout of statistical analyses to get to a final list of DMRs, hence this analysis is only to highlight results which may be of greater importance to the physiological response of actinia equina to their experimental contexts.

# Installing and Loading required packages

```{r check_packages_ClusterProfilerR, message=FALSE, warning=FALSE, results='hide'}
# Build a function which will automatically detect installed packages and install required packages
check_packages_ClusterProfileR <- function(pkg_list) {
  for (pkg in pkg_list) {
    if (!require(pkg, character.only = TRUE)) {
      print(paste(pkg, "is not installed, installing package..."))
      install.packages(pkg) # installs regular R packages
      BiocManager::install(pkg) # installs packages requiring BiocManager
    } else {
      suppressMessages(library(pkg, character.only = TRUE))
      print(paste(pkg, "is installed and loaded"))
    }
  }
}

# Generate a dataframe with a list of packages required

pkg_list_ClusterProfileR<-c("BiocManager", # Have this first to ensure BiocManager packages will be installed
                            "tidyverse", 
                            "tibble", 
                            "stats",
                            "ggplot2", 
                            "dplyr", 
                            "tidyr", 
                            "ggfortify", 
                            "knitr",
                            "clusterProfiler",
                            "AnnotationForge",
                            "enrichplot",
                            "data.table",
                            "stats",
                            "plotly",
                            "ggVennDiagram",
                            "VennDiagram",
                            "limma",
                            "ggridges",
                            "pathview",
                            "KEGGREST",
                            "stringr",
                            "gridExtra",
                            "ggside")

# Run the code
  # NOTE: this code should be run twice, double checking is always good!

# Run Function::
check_packages_ClusterProfileR(pkg_list_ClusterProfileR)
# You will need to answer "no" when asked if you want to update packages, otherwise it will take a while to load

# Double Check everything has loaded:
check_packages_ClusterProfileR(pkg_list_ClusterProfileR)
```

## convertFractionToNumeric

Used to convert gene ratios in final analyses to numerical values for generating figures

```{r convertFractionToNumeric, message=FALSE, warning=FALSE}
convertFractionToNumeric <- function(fraction) {
  fraction_parts <- strsplit(fraction, "/")[[1]]
  numerator <- as.numeric(fraction_parts[1])
  denominator <- as.numeric(fraction_parts[2])
  numeric_value <- numerator / denominator
  return(numeric_value)
}
```

# Importing Data:

```{r import_data}
getwd()
setwd("C:/Users/r02hw22/OneDrive - University of Aberdeen/Documents/GitHub/Methylation_Analyses/Methylation_Analyses/")


# load in the gene lists for experiments 1 & 2;
exp1_DMR_list <- read.table("exp1_DMR_gene_list.txt")
exp1_DMR_list<- as.vector(exp1_DMR_list[,1])
exp2_DMR_list <- read.table("exp2_DMR_gene_list.txt")
exp2_DMR_list<- as.vector(exp2_DMR_list[,1])


# Genome Matching file:
matching_file <- read.table("genome_matching_file.txt", 
                            header = F, 
                            sep = '\t', 
                            stringsAsFactors = FALSE)

names(matching_file)[names(matching_file) == "V13"] <- "ID"

# Reduce list of hits per gene to only the best per gene id:
matching_file_best_hits <- matching_file %>%
  group_by(V1) %>%
  filter(V11 == min(V11)) %>%
  ungroup()

matching_file_best_hits <- unique(matching_file_best_hits)

# Load in uniprot data
BP_MF_Terms <- read.table("UNIPROT_results.txt", 
                          header = T, 
                          sep = '\t', 
                          stringsAsFactors = FALSE)

names(BP_MF_Terms)[names(BP_MF_Terms) == "From"] <- "ID"
names(BP_MF_Terms)[names(BP_MF_Terms) == "Gene.Names"] <- "GENENAMES"
names(BP_MF_Terms)[names(BP_MF_Terms) == "Gene.Ontology..biological.process."] <- "GO:BP"
names(BP_MF_Terms)[names(BP_MF_Terms) == "Gene.Ontology..molecular.function."] <- "GO:MF"

```

## Merge Blastp and GO terms files together

```{r Merge_Files}
# Merge DEG Files:
merged_df <- merge(matching_file_best_hits, BP_MF_Terms, by = "ID")

# Subset Dataframe
merged_df <- subset(merged_df[, c("V1","ID","GO:BP","GO:MF")])

# Change Variable names
names(merged_df)[names(merged_df) == "V1"] <- "GENE"

```

## Separate BP and MF GOs

```{r Separate_GOs}
# Select BP column:
BP <- merged_df[,c("GENE","ID","GO:BP")]
BP <- BP[!apply(BP == "", 1, any), ]
BP <- BP[!duplicated(BP), ]
# Extract GO terms:
GENES_BP <- BP
GENES_BP$GO <- ""
bracketed_rows <- grepl("\\[.*\\]", GENES_BP$`GO:BP`)
GENES_BP$GO[bracketed_rows] <- gsub(".*\\[(.*)\\].*", "\\1", GENES_BP$`GO:BP`[bracketed_rows])
GENES_BP <- GENES_BP[, c("GO","GENE")]
GENES_BP <- GENES_BP[!duplicated(GENES_BP), ]
GENES_BP <- GENES_BP[complete.cases(GENES_BP),]

## Extract Terms:
TERMs_BP <- BP
TERMs_BP$GO <- ""
bracketed_rows <- grepl("\\[.*\\]", TERMs_BP$`GO:BP`)
TERMs_BP$GO[bracketed_rows] <- gsub(".*\\[(.*)\\].*", "\\1", TERMs_BP$`GO:BP`[bracketed_rows])
TERMs_BP$`GO:BP`[bracketed_rows] <- gsub("\\[.*\\]", "", TERMs_BP$`GO:BP`[bracketed_rows])
TERMs_BP <- TERMs_BP[, c("GO","GO:BP")]
TERMs_BP <- TERMs_BP[!duplicated(TERMs_BP), ]
names(TERMs_BP)[names(TERMs_BP) == "GO:BP"] <- "GENENAMES_BP"
TERMs_BP <- TERMs_BP[complete.cases(TERMs_BP),]

# Select MF column:
MF <- merged_df[,c("GENE","ID","GO:MF")]
MF <- MF[!apply(MF == "", 1, any), ]
MF <- MF[!duplicated(MF), ]
# Extract GO terms:
GENES_MF <- MF
GENES_MF$GO <- ""
bracketed_rows <- grepl("\\[.*\\]", GENES_MF$`GO:MF`)
GENES_MF$GO[bracketed_rows] <- gsub(".*\\[(.*)\\].*", "\\1", GENES_MF$`GO:MF`[bracketed_rows])
GENES_MF <- GENES_MF[, c("GO","GENE")]
GENES_MF <- GENES_MF[!duplicated(GENES_MF), ]
GENES_MF <- GENES_MF[complete.cases(GENES_MF),]

## Extract Terms:
TERMs_MF <- MF
TERMs_MF$GO <- ""
bracketed_rows <- grepl("\\[.*\\]", TERMs_MF$`GO:MF`)
TERMs_MF$GO[bracketed_rows] <- gsub(".*\\[(.*)\\].*", "\\1", TERMs_MF$`GO:MF`[bracketed_rows])
TERMs_MF$`GO:MF`[bracketed_rows] <- gsub("\\[.*\\]", "", TERMs_MF$`GO:MF`[bracketed_rows])
TERMs_MF <- TERMs_MF[, c("GO","GO:MF")]
TERMs_MF <- TERMs_MF[!duplicated(TERMs_MF), ]
names(TERMs_MF)[names(TERMs_MF) == "GO:MF"] <- "GENENAMES_MF"
TERMs_MF <- TERMs_MF[complete.cases(TERMs_MF),]
```


```{r ORA_exp1, results='hide'}
# Run ORA
set.seed(100)
ORA_exp1_BP <- enricher(exp1_DMR_list, 
                TERM2GENE = GENES_BP,
                TERM2NAME = TERMs_BP,
                minGSSize=1,
                pvalueCutoff = 1,
                qvalueCutoff = 1)

set.seed(100)
ORA_exp1_MF <- enricher(exp1_DMR_list, 
                TERM2GENE = GENES_MF,
                TERM2NAME = TERMs_MF,
                minGSSize=1,
                pvalueCutoff = 1,
                qvalueCutoff = 1)

# Extract results and add in GO identifiers
ORA_exp1_BP_results <- ORA_exp1_BP@result
ORA_exp1_BP_results$GO <- "Biological Processes"

ORA_exp1_MF_results <- ORA_exp1_MF@result
ORA_exp1_MF_results$GO <- "Molecular Functions"

# combine results and subset for significant values
ORA_exp1_results <- rbind(ORA_exp1_BP_results, ORA_exp1_MF_results)
# ORA_exp1_results <- subset(ORA_exp1_results[ORA_exp1_results$p.adjust < 0.1,])

# Add column to id this as exp1;
ORA_exp1_results$experiment <- "Acute Experiment"

# Change fractions to numeric values
ORA_exp1_results$GeneRatio <- sapply(ORA_exp1_results$GeneRatio, convertFractionToNumeric)
ORA_exp1_results$BgRatio <- sapply(ORA_exp1_results$BgRatio, convertFractionToNumeric)

## Export Dataframe results
write.csv(ORA_exp1_results, "ORA_exp1_DMR_results.csv", row.names = FALSE)
```

## ORA results

```{r ORA_exp1_results, fig.height = 20, fig.width=15}
# Merged_ORA_exp1_BP$BgRatio <- sapply(merged_ORA_exp1_BP$BgRatio, convertFractionToNumeric)
ORA_exp1_results <- ORA_exp1_results[order(ORA_exp1_results$GeneRatio), ]


ORA_exp1_dotplot <- ggplot(data = ORA_exp1_results, 
                              aes(y = reorder(Description, 
                                              Count, 
                                              increasing = T), 
                                  x = GeneRatio, 
                                  color = p.adjust)) + 
                       geom_point(aes(size = Count), 
                                  stroke = 4) +
                       labs(title = "Biological Processes", y = "GO Description", x = "Gene Ratio") +
                       theme_light() + 
                       scale_color_gradient(low = "red", high = "blue") + 
                       scale_size_continuous(breaks = seq(min(ORA_exp1_results$Count), max(ORA_exp1_results$Count), by = 1)) +
                       theme(
                         legend.spacing = unit(0.5, "cm"),         # Adjust overall legend spacing
                         legend.key.height = unit(3, "lines"), # Adjust the height between size examples
                         axis.text.y = element_text(size = 11)
                         ) + 
                       guides(size = guide_legend(title = "Count"))  + 
                       scale_y_discrete(labels = function(x) str_wrap(x, width = 28)) + # Adjust width for line breaks 
                       facet_wrap(~GO, scales = "free")

# Merge ORA_exp1_BP results and dotplot together, and rm dotplot
plot(ORA_exp1_dotplot)

# Save current Environemnt:
save.image(file = "GSEA_exp1_prep.RData")
```

```{r GSEA_exp1, results = 'hide'}
# import dataframes
exp1_DM_region_data <- read.table("./exp1_DMRs.txt", header = T)
exp2_DM_region_data <- read.table("./exp2_DMRs.txt", header = T)

# Generate GeneList Dataframe:
# Feature 1: numeric vector (meth.diff)
geneList_1 = exp1_DM_region_data[,7]
geneList_2 = exp2_DM_region_data[,7]

# Feature 2: Named Vector (rownames)
names(geneList_1) = as.character(exp1_DM_region_data[,19])
names(geneList_2) = as.character(exp2_DM_region_data[,19])

# feature 3: decreasing order;
geneList_1 = sort(geneList_1, decreasing = TRUE)
geneList_2 = sort(geneList_2, decreasing = TRUE)

# Run GSEA exp1
set.seed(100)
GSEA_exp1_BP<- GSEA(geneList_1,
           TERM2GENE = GENES_BP,
           TERM2NAME = TERMs_BP,
           pvalueCutoff = 0.1)
set.seed(100)
GSEA_exp1_MF<- GSEA(geneList_1,
           TERM2GENE = GENES_MF,
           TERM2NAME = TERMs_MF,
           pvalueCutoff = 0.1)

# Run GSEA exp2
set.seed(100)
GSEA_exp2_BP<- GSEA(geneList_2,
           TERM2GENE = GENES_BP,
           TERM2NAME = TERMs_BP,
           pvalueCutoff = 0.1)
set.seed(100)
GSEA_exp2_MF<- GSEA(geneList_2,
           TERM2GENE = GENES_MF,
           TERM2NAME = TERMs_MF,
           pvalueCutoff = 0.1)

# Get Results exp1
GSEA_exp1_BP_results <- GSEA_exp1_BP@result
GSEA_exp1_BP_results$GO <- "biological_processes"

GSEA_exp1_MF_results <- GSEA_exp1_MF@result
GSEA_exp1_MF_results$GO <- "molecular_function"

GSEA_exp1_results <- rbind(GSEA_exp1_BP_results, GSEA_exp1_MF_results)

# Subset for only significant results
GSEA_exp1_results <- subset(GSEA_exp1_results[GSEA_exp1_results$p.adjust < 0.1,])


# Get Results exp2
GSEA_exp2_BP_results <- GSEA_exp2_BP@result
GSEA_exp2_BP_results$GO <- "biological_processes"

GSEA_exp2_MF_results <- GSEA_exp2_MF@result
GSEA_exp2_MF_results$GO <- "molecular_function"

GSEA_exp2_results <- rbind(GSEA_exp2_BP_results, GSEA_exp2_MF_results)

GSEA_exp2_results <- subset(GSEA_exp2_results[GSEA_exp2_results$p.adjust < 0.1,])


# Make function to capture methylation data for each gene in the GSEA results
get_methDiff_data <- function(core_enrichment, meth_data) {
  genes <- unlist(strsplit(core_enrichment, "/"))
  lfc_values <- meth_data$meth.diff[match(genes, meth_data$gene_id_labelling)]
  paste(lfc_values, collapse = "/")
}

GSEA_exp1_results_with_methDiff <- GSEA_exp1_results %>%
  rowwise() %>%
  mutate(methDiff_combined = get_methDiff_data(core_enrichment, exp1_DM_region_data))

GSEA_exp2_results_with_methDiff <- GSEA_exp2_results %>%
  rowwise() %>%
  mutate(methDiff_combined = get_methDiff_data(core_enrichment, exp2_DM_region_data))

# Working on figures
GSEA_exp1_results_with_methDiff$Gene_IDs_list <- strsplit(as.character(GSEA_exp1_results_with_methDiff$core_enrichment), "/")
GSEA_exp1_results_with_methDiff$MethDiff_list <- strsplit(as.character(GSEA_exp1_results_with_methDiff$methDiff_combined), "/")

expanded_data_exp1 <- GSEA_exp1_results_with_methDiff %>% tidyr::unnest(c(Gene_IDs_list, MethDiff_list))
expanded_data_exp1$MethDiff_list <- as.numeric(expanded_data_exp1$MethDiff_list)

data_unique_exp1 <- expanded_data_exp1 %>% distinct()
GO_data_Exp1 <- data_unique_exp1

GO_data_Exp1$WrappedDescription <- str_wrap(GO_data_Exp1$Description, width = 20)  # Adjust the width as needed

GSEA_exp1_plot <- ggplot(GO_data_Exp1, aes(
  x = reorder(WrappedDescription, MethDiff_list), 
  y = MethDiff_list
)) +
  geom_half_boxplot(
    aes(fill = enrichmentScore,
        group = interaction(WrappedDescription, ID)),
    side = "l",   # left half
    alpha = 0.8
  ) +
  geom_half_point(
    aes(colour = enrichmentScore,
        group = interaction(WrappedDescription, ID)),
    side = "r",   # right half
    alpha = 0.4,
    size = 2
  ) +
  coord_flip() + 
  theme_classic()+
  facet_wrap(~GO, scales = "free" ,
             labeller = as_labeller(
               c(
                 "biological_processes" = "Biological Processes",
                 "molecular_function" = "Molecular Functions")
             )
  ) +
 scale_fill_gradient(low = "#2396de", high = "#db183f" , limits = c(-1,1))+
 scale_colour_gradient(low = "#2396de", high = "#db183f", limits = c(-1,1), 
                       guide = "none") +
  geom_hline(yintercept = 0,  linetype = "dashed",  color = "black", linewidth = 0.7) +
  labs(
    y = "Methylation Difference (%)", 
    x = "Gene Ontology (GO)", 
    fill = "Enrichment Score"
  )

print(GSEA_exp1_plot)



GSEA_exp2_results_with_methDiff$Gene_IDs_list <- strsplit(as.character(GSEA_exp2_results_with_methDiff$core_enrichment), "/")
GSEA_exp2_results_with_methDiff$MethDiff_list <- strsplit(as.character(GSEA_exp2_results_with_methDiff$methDiff_combined), "/")

expanded_data_exp2 <- GSEA_exp2_results_with_methDiff %>% tidyr::unnest(c(Gene_IDs_list, MethDiff_list))
expanded_data_exp2$MethDiff_list <- as.numeric(expanded_data_exp2$MethDiff_list)

data_unique_exp2 <- expanded_data_exp2 %>% distinct()
GO_data_Exp2 <- data_unique_exp2

GO_data_Exp2$WrappedDescription <- str_wrap(GO_data_Exp2$Description, width = 20)  # Adjust the width as needed

GSEA_exp1_plot <- ggplot(GO_data_Exp2, aes(
  x = reorder(WrappedDescription, MethDiff_list), 
  y = MethDiff_list
)) +
  geom_half_boxplot(
    aes(fill = enrichmentScore,
        group = interaction(WrappedDescription, ID)),
    side = "l",   # left half
    alpha = 0.8
  ) +
  geom_half_point(
    aes(colour = enrichmentScore,
        group = interaction(WrappedDescription, ID)),
    side = "r",   # right half
    alpha = 0.4,
    size = 2
  ) +
  coord_flip() + 
  theme_classic()+
  facet_wrap(~GO, scales = "free" ,
             labeller = as_labeller(
               c(
                 "biological_processes" = "Biological Processes",
                 "molecular_function" = "Molecular Functions")
             )
  ) +
 scale_fill_gradient(low = "#2396de", high = "#db183f" , limits = c(-1,1))+
 scale_colour_gradient(low = "#2396de", high = "#db183f", limits = c(-1,1), 
                       guide = "none") +
  geom_hline(yintercept = 0,  linetype = "dashed",  color = "black", linewidth = 0.7) +
  labs(
    y = "Methylation Difference (%)", 
    x = "Gene Ontology (GO)", 
    fill = "Enrichment Score"
  )

print(GSEA_exp1_plot)

```

```{r ORA_exp2, results='hide'}
# Run ORA
set.seed(100)
ORA_exp2_BP <- enricher(exp2_DMR_list, 
                TERM2GENE = GENES_BP,
                TERM2NAME = TERMs_BP,
                minGSSize=1,
                pvalueCutoff = 1,
                qvalueCutoff = 1)

set.seed(100)
ORA_exp2_MF <- enricher(exp2_DMR_list, 
                TERM2GENE = GENES_MF,
                TERM2NAME = TERMs_MF,
                minGSSize=1,
                pvalueCutoff = 1,
                qvalueCutoff = 1)

# Extract results and add in GO identifiers
ORA_exp2_BP_results <- ORA_exp2_BP@result
ORA_exp2_BP_results$GO <- "Biological Processes"

ORA_exp2_MF_results <- ORA_exp2_MF@result
ORA_exp2_MF_results$GO <- "Molecular Functions"

# combine results and subset for significant values
ORA_exp2_results <- rbind(ORA_exp2_BP_results, ORA_exp2_MF_results)
# ORA_exp2_results <- subset(ORA_exp2_results[ORA_exp2_results$p.adjust < 0.1,])

# Adding experiment column to differentiate between experiemnts for plotting
ORA_exp2_results$experiment <- "Primed Experiment"

# Change fractions to numeric values
ORA_exp2_results$GeneRatio <- sapply(ORA_exp2_results$GeneRatio, convertFractionToNumeric)
ORA_exp2_results$BgRatio <- sapply(ORA_exp2_results$BgRatio, convertFractionToNumeric)

## Export Dataframe results
write.csv(ORA_exp2_results, "ORA_exp2_DMR_results.csv", row.names = FALSE)
```

## ORA results

```{r ORA_exp2_results, fig.height = 8, fig.width=12}
# Merged_ORA_exp2_BP$BgRatio <- sapply(merged_ORA_exp2_BP$BgRatio, convertFractionToNumeric)
ORA_exp2_results <- ORA_exp2_results[order(ORA_exp2_results$GeneRatio), ]


ORA_exp2_dotplot <- ggplot(data = ORA_exp2_results, 
                              aes(y = reorder(Description, 
                                              Count, 
                                              increasing = T), 
                                  x = GeneRatio, 
                                  color = p.adjust)) + 
                       geom_point(aes(size = Count), 
                                  stroke = 4) +
                       labs(y = "GO Description", x = "Gene Ratio") +
                       theme_light() + 
                       scale_color_gradient(low = "red", high = "blue") + 
                       scale_size_continuous(breaks = seq(min(ORA_exp2_results$Count), max(ORA_exp2_results$Count), by = 1)) +
                       theme(
                         legend.spacing = unit(0.5, "cm"),         # Adjust overall legend spacing
                         legend.key.height = unit(3, "lines"), # Adjust the height between size examples
                         axis.text.y = element_text(size = 10)
                         ) + 
                       guides(size = guide_legend(title = "Count"))  + 
                       scale_y_discrete(labels = function(x) str_wrap(x, width = 20)) + # Adjust width for line breaks 
                       facet_wrap(~GO, scales = "free")

plot(ORA_exp2_dotplot)


# Save current Environemnt:
save.image(file = "GSEA_exp2_prep.RData")
```

```{r combining_results}
ORA_exp1_results <- ORA_exp1_results[order(ORA_exp1_results$GeneRatio), ]
ORA_exp2_results <- ORA_exp2_results[order(ORA_exp2_results$GeneRatio), ]

ORA_results_combined <- rbind(ORA_exp1_results, ORA_exp2_results)


ORA_results_combined_dotplot <- ggplot(data = ORA_results_combined, 
                              aes(y = reorder(Description, 
                                              Count, 
                                              increasing = T), 
                                  x = GeneRatio, 
                                  color = p.adjust)) + 
                       geom_point(aes(size = Count), 
                                  stroke = 2.5) +
                       labs(y = "GO Description", x = "Gene Ratio") +
                       theme_light() + 
                       scale_color_gradient(low = "red", high = "blue") + 
                       scale_size_continuous(breaks = seq(min(ORA_results_combined$Count), max(ORA_results_combined$Count), by = 1)) +
                       theme(
                         legend.spacing = unit(0.5, "cm"),         # Adjust overall legend spacing
                         legend.key.height = unit(3, "lines"), # Adjust the height between size examples
                         axis.text.y = element_text(size = 10)
                         ) + 
                       guides(size = guide_legend(title = "Count"))  + 
                       scale_y_discrete(labels = function(x) str_wrap(x, width = 20)) + # Adjust width for line breaks 
                       facet_wrap(~experiment, scales = "free")

plot(ORA_results_combined_dotplot)

```



```{r}
matched_exp1_bp <- merge(
  data.frame(GENE = exp1_DMR_list),
  GENES_BP,
  by.x = "GENE",
  #by.y = "gene_id",
  all.x = FALSE,
  all.y = FALSE
)

write.csv(matched_exp1_bp, "matched_exp1_bp.csv", row.names = FALSE)

matched_exp1_bp$description <- c("response to stimulus",
                                 "innate immune response",
                                 "DNA repair",
                                 "regulation of stem cell population maintenance")

matched_exp1_bp$ontology <- "Biological Processes"
matched_exp1_bp$experiment <- "Acute Experiment"

# matched_exp1_bp_terms <- merge(
#   matched_exp1_bp,
#   TERMs_BP,
#   by.x = "GO",
#   all.x = F,
#   all.y = F
# )

matched_exp1_mf <- merge(
  data.frame(GENE = exp1_DMR_list),
  GENES_MF,
  by.x = "GENE",
  #by.y = "gene_id",
  all.x = FALSE,
  all.y = FALSE
)

matched_exp1_mf$description <- c("RNA helicase activity",
                                  "RNA binding",
                                  "zinc ion binding",
                                  "zinc ion binding",
                                  "RNA binding")

matched_exp1_mf$ontology <- "Molecular Functions"
matched_exp1_mf$experiment <- "Acute Experiment"

# matched_exp1_mf_terms <- merge(
#   matched_exp1_mf,
#   TERMs_MF,
#   by.x = "GO",
#   all.x = F,
#   all.y = F
# )

matched_exp2_bp <- merge(
  data.frame(GENE = exp2_DMR_list),
  GENES_BP,
  by.x = "GENE",
  #by.y = "gene_id",
  all.x = FALSE,
  all.y = FALSE
)

matched_exp2_bp$description <- c("lipid oxidation",
                                 "lipid oxidation",
                                 "nitrogen compound transport",
                                 "nitrogen compound transport",
                                 "DNA integration",
                                 "DNA integration",
                                 "carbohydrate metabolic process")

matched_exp2_bp$ontology <- "Biological Processes"
matched_exp2_bp$experiment <- "Primed Experiment"

matched_exp2_mf <- merge(
  data.frame(GENE = exp2_DMR_list),
  GENES_MF,
  by.x = "GENE",
  #by.y = "gene_id",
  all.x = FALSE,
  all.y = FALSE
)

matched_exp2_mf$description <- c("metal ion binding",
                                 "oxidoreductase activity, acting on single donors with incorporation of molecular oxygen, incorporation of two atoms of oxygen",
                                 "metal ion binding",
                                 "oxidoreductase activity, acting on single donors with incorporation of molecular oxygen, incorporation of two atoms of oxygen",
                                 "metal ion binding",
                                 "transmembrane transport activity",
                                 "transmembrane transport activity",
                                 "zinc ion binding",
                                 "nucleic acid binding",
                                 "polysaccharide binding")

matched_exp2_mf$ontology <- "Molecular Functions"
matched_exp2_mf$experiment <- "Primed Experiment"

all_matched_data <- rbind(matched_exp1_bp, matched_exp1_mf, matched_exp2_bp, matched_exp2_mf)

all_matched_data

```



# Calculating and comparing proportions of genome aganist DMR features:

```{r}

exp1_DMRs <- read.csv("exp1_DMRs_refined.csv", header = T)
exp1_DMRs <- subset(exp1_DMRs[exp1_DMRs$p_fdr <= 0.1,])

exp1_DMRs$region[exp1_DMRs$region == "three_prime_UTR"] <- "3` UTR"


exp2_DMRs <- read.table("exp2_DMRs.txt", header = T, sep = "\t")
exp2_DMRs <- subset(exp2_DMRs[exp2_DMRs$p_fdr <= 0.1,])

# If your columns are named differently, adjust these selections accordingly
df_summary_exp1 <- exp1_DMRs %>%
  group_by(region) %>%                # Adjust "feature" if your column has a different name
  summarize(feature_count = n()) %>%   # Count how many rows per feature
  mutate(proportion_percent = 100 * feature_count / sum(feature_count))

# If your columns are named differently, adjust these selections accordingly
df_summary_exp2 <- exp2_DMRs %>%
  group_by(region) %>%                # Adjust "feature" if your column has a different name
  summarize(feature_count = n()) %>%   # Count how many rows per feature
  mutate(proportion_percent = 100 * feature_count / sum(feature_count))

# View the summarized data
print(df_summary_exp1)
print(df_summary_exp2)

# Construct the summary of the combined_annotations.gff3 file:
combined_annotations_summary <- NULL

combined_annotations_summary$region <- c("Repeat",
                                         "Downstream",
                                         "exon",
                                         "5` UTR",
                                         "Gene",
                                          "ncRNA",
                                          "Promoter",
                                          "3` UTR")

combined_annotations_summary$feature_count <- c(978221,
                                                47671,
                                                383432,
                                                23401,
                                                47671,
                                                7980,
                                                47671,
                                                23356
                                                )



combined_annotations_summary$proportion_percent <- c(62.73,
                                                     3.06,
                                                     24.59,
                                                     1.50,
                                                     3.06,
                                                     0.51,
                                                     3.06,
                                                     1.50
                                                     )

combined_annotations_summary <- as.data.frame(combined_annotations_summary)

library(ggplot2)

# Sort from largest to smallest so that
# the largest portion is drawn first (on the bottom of the bar).
# df_summary <- df_summary %>%
#   arrange(proportion_percent)
# 
# # Turn 'region' into a factor with levels in descending order
# df_summary$region <- factor(df_summary$region, levels = df_summary$region)
# 
# # Bar plot by proportion
# ggplot(df_summary, aes(x = region, y = proportion_percent, fill = region)) +
#   geom_bar(stat = "identity") +
#   labs(title = "Bar Plot of Feature Proportions",
#        x = "Feature",
#        y = "Proportion (%)") +
#   theme_bw()
# 
# 
# # Sort from largest to smallest so that
# # the largest portion is drawn first (on the bottom of the bar).
# combined_annotations_summary <- combined_annotations_summary %>%
#   arrange(proportion_percent)
# 
# # Turn 'region' into a factor with levels in descending order
# combined_annotations_summary$region <- factor(combined_annotations_summary$region, levels = combined_annotations_summary$region)
# 
# # Bar plot by proportion
# ggplot(combined_annotations_summary, aes(x = region, y = proportion_percent, fill = region)) +
#   geom_bar(stat = "identity") +
#   labs(
#     title = "Bar Plot of Feature Proportions",
#     x = "Feature",
#     y = "Proportion (%)"
#   ) +
#   theme_bw() +
#   theme(
#     axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)
#   )



df_summary_exp1$type <- "DMRs - Acute"
df_summary_exp2$type <- "DMRs - Primed"
combined_annotations_summary$type <- "Annotated Genome"

all_data <- rbind(df_summary_exp1, df_summary_exp2, combined_annotations_summary)

all_data <- all_data %>%
  arrange(proportion_percent)

all_data$region <- as.factor(all_data$region)

# rename values to be correct
all_data$region[all_data$region == "dispersed_repeat"] <- "Repeat"
all_data$region[all_data$region == "promoter"] <- "Promoter"

# Set the order of feature_type as a factor
all_data$region <- factor(all_data$region, levels = c("Repeat", 
                                                                  "exon",
                                                                  "Promoter",
                                                                  "Gene",
                                                                  "Downstream",
                                                                  "3` UTR",
                                                                  "5` UTR",
                                                                  "ncRNA"))



rel_plot <- ggplot(all_data, aes(x = region, y = proportion_percent, fill = region)) +
  geom_bar(stat = "identity") +
  geom_text(aes(label = feature_count), vjust = -0.5, color = "black", angle = 45, hjust = 0)+
  labs(
    x = "Feature",
    y = "Proportion (%)"
  ) +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)
  ) + 
  facet_wrap(~type, nrow = 1, scales = "free_x") +
    scale_fill_manual(
    name = "Region",
    values = c("#E69F00", "#56B4E9", "#009E73", "#a1d61a", "#0072B2", "#D55E00", "#CC79A7", "red"),
    labels = c("Repeat", "Downstream", "Exon", "5' UTR", "Gene","ncRNA", "Promoter", "3' UTR")) 

rel_plot

ggsave("./rel_plot.png", rel_plot, width = 30, height = 20, units= "cm", dpi = 600)


abs_plot <- ggplot(all_data, aes(x = region, y = feature_count, fill = region)) +
  geom_bar(stat = "identity") +
  labs(
    x = "Feature",
    y = "Count of Features"
  ) +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1, vjust = 1)
  ) + 
  facet_wrap(~type, nrow = 1, scales = "free")



grid.arrange(abs_plot, rel_plot, nrow = 2)

install.packages("gtable")
library(gtable)

g1 <- ggplotGrob(abs_plot)
g2 <- ggplotGrob(rel_plot)
g <- rbind(g1, g2, size = "first")
g$widths <- unit.pmax(g1$widths, g2$widths)
grid.newpage()
grid.draw(g)


```

# Making a new figure with the uniprot matched GO ids and descriptions of DMRs

```{r}
data <- read.table("DMR functional terms table.txt", header = T, sep = "\t")

data <- data[ !apply(data =="", 1, any),]

data <- data %>%
  arrange(Methylation.Difference....)

data$Methylation.Difference.... <- as.numeric(data$Methylation.Difference....)
data$GO <- as.factor(data$GO)
data$GO.Type <- as.factor(data$GO.Type)
data$Feature.type <- as.factor(data$Feature.type)
data$Experiment <- as.factor(data$Experiment)
data$Cytosine.Context <- as.factor(data$Cytosine.Context)
data$Enriched <- as.factor(data$Enriched)

data$Wrapped_GO <- str_wrap(data$GO, width = 30)  # Adjust the width as needed


cols <- c("#e08282", "#f5e9e9")
cols2 <- c("red2", "#4d4949")

ggplot(
  data = data,
  aes(
    x = Methylation.Difference....,
    y = reorder(Wrapped_GO, Methylation.Difference....),
    color = Enriched,
    fill = Enriched
  )
) +
  geom_boxplot(alpha = 0.5) +
  facet_wrap(~Experiment + GO.Type, scales = "free") +
  labs(
    x = "Methylation Difference (%)",
    y = "Gene Ontology (GO)",
    fill = "Gene Ontology",
    color = "Gene Ontology",
    size = "Adjusted P-Value"
  ) +
  theme_bw() +
  scale_fill_manual(values = cols) +
  scale_colour_manual(values = cols2)

ggsave("ORA_DMRs.png",width = 25, height = 30, dpi = 300, units = "cm")

# Need to change the descriptions of the functionlly enriched terms so that I can highlight them on plot;
ORA_results_combined

write.table(ORA_results_combined, "ORA_results_combined_TO_BE_RENAMED.txt", sep = '\t')

ORA_results_combined_RENAMED <- read.table("ORA_results_combined_RENAMED.txt", sep = '\t', header = T)



# -----------------------------------------------------------------------------

# Making a plot of just the acute exposure expeirment
# including the ggside splits to show the number of parts associated to each function

data_acute <- data[data$Experiment == "Acute",]
data_acute$Feature.type <- as.character(data_acute$Feature.type)
str(data_acute)
data_primed <- data[data$Experiment == "Primed",]

# Change colour parameters here:
cols <- c("#e08282", "#f5e9e9")
cols2 <- c("red2", "#4d4949")

library(ggside)
library(gghalves)

# Acute plot
acute_fe_plot <- ggplot(
  data = data_acute,
  aes(
    x = Methylation.Difference....,
    y = reorder(Wrapped_GO, Methylation.Difference....),
    color = Enriched,
    fill = Enriched
  )
) +
  geom_boxplot(alpha = 0.5) +
  facet_wrap(~GO.Type, scales = "free",
             labeller = as_labeller(
               c(
                 "BP" = "Biological Processes",
                 "MF" = "Molecular Functions")
                )
             ) +
  labs(
    x = "Methylation Difference (%)",
    y = "Gene Ontology (GO)",
    fill = "Overrepresentation",
    color = "Overrepresentation",
    size = "Adjusted P-Value"
  ) +
  theme_bw() +
  scale_fill_manual(values = cols) +
  scale_colour_manual(values = cols2)

acute_fe_plot

ggsave("./acute_fe_plot.png", acute_fe_plot, width = 2, height = 16, units= "cm", dpi = 600)

# Primed plot
primed_fe_plot <- ggplot(
  data = data_primed,
  aes(
    x = Methylation.Difference....,
    y = reorder(Wrapped_GO, Methylation.Difference....),
    color = Enriched,
    fill = Enriched
  )
) +
  geom_boxplot(alpha = 0.5) +
  facet_wrap(~GO.Type, 
             scales = "free",
             labeller = as_labeller(
               c(
                 "BP" = "Biological Processes",
                 "MF" = "Molecular Functions")
                )
             ) +
  labs(
    x = "Methylation Difference (%)",
    y = "Gene Ontology (GO)",
    fill = "Overrepresentation",
    color = "Overrepresentation",
    size = "Adjusted P-Value"
  ) +
  theme_bw() +
  scale_fill_manual(values = cols) +
  scale_colour_manual(values = cols2)

primed_fe_plot

ggsave("./primed_fe_plot.png", primed_fe_plot, width = 24, height = 16, units= "cm", dpi = 600)

getwd()
```

# Correlating Expression and Methylation

```{r}

# Importing data file which already has expression and methylation combined from using the 'VLOOKUP' function in excel, I used the full DMR results for both experiments, alongside the full DEA results for both experiments, homogenised the naming conventions (removing .TU, and replacing .path with .mrna).


getwd()

meth_exp_data <- read.table("Corelating Methylation and Expression All context data.txt", header = T, sep = '\t')

str(meth_exp_data)

meth_exp_data <- meth_exp_data %>%
  mutate(region = ifelse(region == "dispersed_repeat", "Repeat", region))
meth_exp_data <- meth_exp_data %>%
  mutate(region = ifelse(region == "downstream_region", "Downstream", region))
meth_exp_data <- meth_exp_data %>%
  mutate(region = ifelse(region == "exon", "Exon", region))        
meth_exp_data <- meth_exp_data %>%
  mutate(region = ifelse(region == "five_prime_UTR", "5` UTR", region))
meth_exp_data <- meth_exp_data %>%
  mutate(region = ifelse(region == "ncRNA_gene", "ncRNA", region))
meth_exp_data <- meth_exp_data %>%
  mutate(region = ifelse(region == "promoter", "Promoter", region))
meth_exp_data <- meth_exp_data %>%
  mutate(region = ifelse(region == "three_prime_UTR", "3` UTR", region))
meth_exp_data <- meth_exp_data %>%
  mutate(region = ifelse(region == "gene", "Gene", region))
meth_exp_data <- meth_exp_data %>%
  mutate(experiment = ifelse(experiment == "exp1", "Acute Experiment", experiment))
meth_exp_data <- meth_exp_data %>%
  mutate(experiment = ifelse(experiment == "exp2", "Primed Experiment", experiment))


meth_exp_data$experiment <- as.factor(meth_exp_data$experiment)
meth_exp_data$context <- as.factor(meth_exp_data$context)
meth_exp_data$region <- as.factor(meth_exp_data$region)


library(ggplot2)

ggplot(meth_exp_data, 
       aes(x = meth.diff,
           y = exp_l2fc,
           col = region
           )) + 
  geom_point(alpha = 0.5, size = 3) + 
  geom_smooth(method = "lm", se = F, color = "black", lwd = 2) +
  geom_vline(xintercept = c(0), col = "black", linetype = "dashed") +
  geom_hline(yintercept = c(0), col = "black", linetype = "dashed") +
  scale_color_manual(
    name = "Region",
    values = c("#E69F00", "#56B4E9", "#009E73", "#a1d61a", "#0072B2", "#D55E00", "#CC79A7", "red"),
    labels = c("Repeat", "Downstream", "Exon", "5' UTR", "Gene","ncRNA", "Promoter", "3' UTR")) +
  labs(
    x = "Methylation Difference (%)",
    y = "Expression Change (Log2FoldChange)",
    color = "Region Type"
  ) +
  theme_bw() + 
  scale_x_continuous(limits = c(-40, 40), breaks = c(-40,-20,0,20,40)) + 
  scale_y_continuous(limits = c(-5, 5), breaks = c(-5,-4,-3, -2, -1, 0, 1, 2, 3,4,5)) + 
  facet_grid(~experiment~region)

# -----------------------------------------------------------------------------

# Making a nice set of plots

levels(meth_exp_data$region)
# reorder regions
meth_exp_data$region <- factor(
  meth_exp_data$region,
  levels = c("Repeat", "Downstream", "Exon", "5` UTR",
             "Gene", "ncRNA", "Promoter", "3` UTR")  # <- Your desired order
)

exp1_corr_plot <- ggplot(meth_exp_data[(meth_exp_data$experiment == "Acute Experiment") #& (meth_exp_data$meth_p_fdr <= 0.1)
                                       ,], 
       aes(x = meth.diff,
           y = exp_l2fc,
           col = region
           )) + 
  geom_point(alpha = 0.5, size = 2, aes(shape = context)) + 
  geom_smooth(method = "lm", se = F, color = "black", lwd = 1) +
  geom_vline(xintercept = c(0), col = "black", linetype = "dashed") +
  geom_hline(yintercept = c(0), col = "black", linetype = "dashed") +
  scale_color_manual(
    name = "Region",
    values = c("#E69F00", "#56B4E9", "#009E73", "#a1d61a", "#0072B2", "#D55E00", "#CC79A7", "red"),
    labels = c("Repeat", "Downstream", "Exon", "5' UTR", "Gene","ncRNA", "Promoter", "3' UTR")) +
  labs(
    x = "Methylation Difference (%)",
    y = "Expression Change (Log2FoldChange)",
    color = "Region Type"
  ) +
  theme_bw() + 
  #scale_x_continuous(limits = c(-40, 40), breaks = c(-40,-20,0,20,40)) + 
  scale_y_continuous(limits = c(-5, 5), breaks = c(-5,-4,-3, -2, -1, 0, 1, 2, 3,4,5)) + 
  facet_wrap(~region, nrow = 2, scales = "free")  + 
  guides(color = "none")

print(exp1_corr_plot)

ggsave("exp1_meth-expression_plot.png", exp1_corr_plot, width = 18, height = 10, units = "cm")

# Create an empty results dataframe
corr_results_exp1 <- data.frame(
  region = character(),
  correlation = numeric(),
  p_value = numeric(),
  stringsAsFactors = FALSE
)

meth_exp_data_exp1 <- meth_exp_data[meth_exp_data$experiment == "Acute Experiment",]

# Loop through each level of "region"
for (reg in levels(meth_exp_data_exp1$region)) {
  
  # Subset the data for this region
  region_data <- subset(meth_exp_data_exp1, region == reg)
  
  # Perform correlation test
  test <- cor.test(region_data$meth.diff, region_data$exp_l2fc)
  
  # Add the result to the results dataframe
  corr_results_exp1 <- rbind(corr_results_exp1, data.frame(
    region = reg,
    correlation = test$estimate,
    p_value = test$p.value
  ))
}


exp2_corr_plot <- ggplot(meth_exp_data[(meth_exp_data$experiment == "Primed Experiment") # & (meth_exp_data$meth_p_fdr <= 0.1) # also an option to look at only the significantly differentiall expressed regions.
                                       ,], 
       aes(x = meth.diff,
           y = exp_l2fc,
           col = region
           )) + 
  geom_point(alpha = 0.5, size = 2, aes(shape = context)) + 
  geom_smooth(method = "lm", se = F, color = "black", lwd = 1) +
  geom_vline(xintercept = c(0), col = "black", linetype = "dashed") +
  geom_hline(yintercept = c(0), col = "black", linetype = "dashed") +
  scale_color_manual(
    name = "Region",
    values = c("#E69F00", "#56B4E9", "#009E73", "#a1d61a", "#0072B2", "#D55E00", "#CC79A7", "red"),
    labels = c("Repeat", "Downstream", "Exon", "5' UTR", "Gene","ncRNA", "Promoter", "3' UTR")) +
  labs(
    x = "Methylation Difference (%)",
    y = "Expression Change (Log2FoldChange)",
    color = "Region Type"
  ) +
  theme_bw() + 
  #scale_x_continuous(limits = c(-40, 40), breaks = c(-40,-20,0,20,40)) + 
  scale_y_continuous(limits = c(-5, 5), breaks = c(-5,-4,-3, -2, -1, 0, 1, 2, 3,4,5)) + 
  facet_wrap(~region, nrow = 2, scales = "free")  + 
  guides(color = "none")

print(exp2_corr_plot)

ggsave("exp2_meth-expression_plot.png", exp2_corr_plot, width = 18, height = 10, units = "cm")


# Create an empty results dataframe
corr_results_exp2 <- data.frame(
  region = character(),
  correlation = numeric(),
  p_value = numeric(),
  stringsAsFactors = FALSE
)

meth_exp_data_exp2<- meth_exp_data[meth_exp_data$experiment == "Primed Experiment",]

corr_results_exp2 <- NULL

# Loop through each level of "region"
for (reg in levels(meth_exp_data_exp2$region)) {
  
  # Subset the data for this region
  region_data <- subset(meth_exp_data_exp2, region == reg)
  
  # Perform correlation test
  test <- cor.test(region_data$meth.diff, region_data$exp_l2fc)
  
  # Add the result to the results dataframe
  corr_results_exp2 <- rbind(corr_results_exp2, data.frame(
    region = reg,
    correlation = test$estimate,
    p_value = test$p.value
  ))
}


# -------------------------------------------------------------------------------

exp1_meth_exp_data <- meth_exp_data[meth_exp_data$experiment == "Acute Experiment",]
exp1_corr_results <- cor.test(exp1_meth_exp_data$meth.diff, 
                              exp1_meth_exp_data$exp.l2fc, 
                              method = "pearson")

print(exp1_corr_results)
  # p = 0.003123
  # correlation = -0.03447973 

summary(exp1_meth_exp_data)

exp2_meth_exp_data <- meth_exp_data[meth_exp_data$experiment == "Primed Experiment",]
exp2_corr_results <- cor.test(exp2_meth_exp_data$meth.diff, 
                              exp2_meth_exp_data$l2fc, 
                              method = "pearson")

print(exp2_corr_results)
  # p = 0.7574
  # correlation = -0.003096255 

summary(exp2_meth_exp_data)


lm <- lm(data = meth_exp_data,
         l2fc~meth.diff)
summary(lm)
anova(lm)

lm_exp1 <- lm(data = meth_exp_data[meth_exp_data$experiment == "Acute Experiment",],
         l2fc~meth.diff+region)
summary(lm_exp1)
anova(lm_exp1)

lm_exp2 <- lm(data = meth_exp_data[meth_exp_data$experiment == "Primed Experiment",],
         l2fc~meth.diff+region)
summary(lm_exp2)
anova(lm_exp2)


```



```{r}


```